# Automatically generated by openapi-generator (https://openapi-generator.tech)
# Please update as you see appropriate

context("Test GeneApi")

api_instance <- GeneApi$new()

test_that("DownloadGenePackage", {
  # tests for DownloadGenePackage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a gene dataset by gene ID
  # Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by gene ID.
  # @param gene_ids array[integer] NCBI gene ids
  # @param include_annotation_type array[V2Fasta] Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (optional)
  # @param fasta_filter array[character] Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated) (optional)
  # @param accession_filter array[character] Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions (optional)
  # @param aux_report array[V2GeneDatasetRequestGeneDatasetReportType] list additional reports to include with download. Data report is included by default. (optional)
  # @param tabular_reports array[V2GeneDatasetRequestGeneDatasetReportType]  (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead. (optional)
  # @param table_report_type V2GeneDatasetRequestGeneDatasetReportType Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead. (optional)
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadGenePackagePost", {
  # tests for DownloadGenePackagePost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a gene dataset by POST
  # Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by POST.
  # @param v2_gene_dataset_request V2GeneDatasetRequest 
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneChromosomeSummary", {
  # tests for GeneChromosomeSummary
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get summary of chromosomes for a particular taxon&#39;s annotation
  # Get summary of chromosomes for a particular taxon&#39;s annotation.
  # @param taxon character 
  # @param annotation_name character 
  # @return [V2GeneChromosomeSummaryReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneCountsForTaxon", {
  # tests for GeneCountsForTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene counts by taxonomic identifier
  # Get gene counts for a specified NCBI Taxonomy ID or name (common or scientific) in JSON format.
  # @param taxon character Taxon for provided gene symbol
  # @return [V2GeneCountsByTaxonReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneCountsForTaxonByPost", {
  # tests for GeneCountsForTaxonByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene counts by taxonomic identifier
  # Get gene counts for a specified NCBI Taxonomy ID or name (common or scientific) in JSON format.
  # @param v2_gene_counts_by_taxon_request V2GeneCountsByTaxonRequest 
  # @return [V2GeneCountsByTaxonReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDatasetReport", {
  # tests for GeneDatasetReport
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene dataset report as JSON
  # Get gene dataset report. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param v2_gene_dataset_reports_request V2GeneDatasetReportsRequest 
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDatasetReportByAccession", {
  # tests for GeneDatasetReportByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by accession IDs
  # Get dataset reports by accession IDs.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param accessions array[character] RNA or Protein accessions.
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDatasetReportByTaxAndSymbol", {
  # tests for GeneDatasetReportByTaxAndSymbol
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by taxons.
  # Get dataset reports by taxons.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param symbols array[character] Gene symbol
  # @param taxon character Taxon for provided gene symbol
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDatasetReportsById", {
  # tests for GeneDatasetReportsById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by gene IDs.
  # Get dataset reports by gene IDs.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param gene_ids array[integer] NCBI gene ids
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDatasetReportsByLocusTag", {
  # tests for GeneDatasetReportsByLocusTag
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene dataset reports by locus tag
  # Get a gene summary for one or more gene locus tags. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param locus_tags array[character] Gene locus tags
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDatasetReportsByTaxon", {
  # tests for GeneDatasetReportsByTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene dataset reports by taxonomic identifier
  # Get a gene summary for a specified NCBI Taxonomy ID or name (common or scientific). By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param taxon character NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDownloadSummaryById", {
  # tests for GeneDownloadSummaryById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene download summary by GeneID
  # Get a download summary by GeneID in a JSON output format.
  # @param gene_ids array[integer] NCBI gene ids
  # @param include_annotation_type array[V2Fasta] Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (optional)
  # @param returned_content V2GeneDatasetRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param fasta_filter array[character] Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated) (optional)
  # @param accession_filter array[character] Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions (optional)
  # @param aux_report array[V2GeneDatasetRequestGeneDatasetReportType] list additional reports to include with download. Data report is included by default. (optional)
  # @param tabular_reports array[V2GeneDatasetRequestGeneDatasetReportType]  (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead. (optional)
  # @param table_report_type V2GeneDatasetRequestGeneDatasetReportType Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead. (optional)
  # @return [V2DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDownloadSummaryByPost", {
  # tests for GeneDownloadSummaryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene download summary
  # Get gene download summary in a JSON output format.
  # @param v2_gene_dataset_request V2GeneDatasetRequest 
  # @return [V2DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneLinksById", {
  # tests for GeneLinksById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene links by gene ID
  # Get links to available gene resources by gene ID.
  # @param gene_ids array[integer] NCBI gene ids, limited to 1000 ids
  # @return [V2GeneLinksReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneLinksByIdByPost", {
  # tests for GeneLinksByIdByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene links by gene ID
  # Get links to available gene resources by gene ID.
  # @param v2_gene_links_request V2GeneLinksRequest 
  # @return [V2GeneLinksReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataByAccession", {
  # tests for GeneMetadataByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene metadata by RefSeq Accession
  # Get a gene summary by RefSeq Accession. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param accessions array[character] RNA or Protein accessions.
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param locus_tags array[character] Gene locus tags (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataByPost", {
  # tests for GeneMetadataByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene metadata as JSON
  # Get a gene summary. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param v2_gene_dataset_reports_request V2GeneDatasetReportsRequest 
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataByTaxAndSymbol", {
  # tests for GeneMetadataByTaxAndSymbol
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene metadata by gene symbol
  # Get a gene summary by by gene symbol. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param symbols array[character] Gene symbol
  # @param taxon character Taxon for provided gene symbol
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param locus_tags array[character] Gene locus tags (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneOrthologsById", {
  # tests for GeneOrthologsById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene orthologs by gene ID
  # Get a gene summary for an ortholog set by gene ID in a JSON output format.
  # @param gene_id integer 
  # @param returned_content V2OrthologRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param taxon_filter array[character] Filter genes by taxa (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `OrthologRequest` call with more than `page_size` results. Use this token, along with the previous `OrthologRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneOrthologsByPost", {
  # tests for GeneOrthologsByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene orthologs by gene ID
  # Get a gene summary for an ortholog set by gene ID in a JSON output format by POST
  # @param v2_ortholog_request V2OrthologRequest 
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneProductReport", {
  # tests for GeneProductReport
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene product reports as JSON
  # Get gene product reports. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param v2_gene_dataset_reports_request V2GeneDatasetReportsRequest 
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneProductReportByAccession", {
  # tests for GeneProductReportByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene product reports by accession IDs
  # Get gene product reports by accession IDs.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param accessions array[character] RNA or Protein accessions.
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneProductReportByTaxAndSymbol", {
  # tests for GeneProductReportByTaxAndSymbol
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get product reports by taxon.
  # Get gene product reports by taxon.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param symbols array[character] Gene symbol
  # @param taxon character Taxon for provided gene symbol
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneProductReportsById", {
  # tests for GeneProductReportsById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene product reports by gene IDs.
  # Get gene product reports by gene IDs.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param gene_ids array[integer] NCBI gene ids
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneProductReportsByLocusTags", {
  # tests for GeneProductReportsByLocusTags
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene product reports by locus tags
  # Get gene product reports for one or more gene locus tags. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param locus_tags array[character] Gene locus tags
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneProductReportsByTaxon", {
  # tests for GeneProductReportsByTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene product reports by taxonomic identifier
  # Get gene product reports for a specified NCBI Taxonomy ID or name (common or scientific). By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param taxon character NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format character Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneReportsById", {
  # tests for GeneReportsById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene reports by GeneID
  # Get a gene summary by GeneID. By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param gene_ids array[integer] NCBI gene ids
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param locus_tags array[character] Gene locus tags (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneReportsByTaxon", {
  # tests for GeneReportsByTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get gene reports by taxonomic identifier
  # Get a gene summary for a specified NCBI Taxonomy ID or name (common or scientific). By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param taxon character NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
  # @param returned_content V2GeneDatasetReportsRequestContentType Return either gene-ids, or entire gene metadata (optional)
  # @param locus_tags array[character] Gene locus tags (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_size integer The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param query character text search within gene symbol, aliases, name, locus-tag and protein name (optional)
  # @param types array[V2GeneType] Gene types to filter (optional)
  # @param accession_filter array[character] Filter tabular product report to only include the selected accessions (optional)
  # @param tax_search_subtree character For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2reportsGeneDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})
