# Automatically generated by openapi-generator (https://openapi-generator.tech)
# Please update as you see appropriate

context("Test GenomeApi")

api_instance <- GenomeApi$new()

test_that("AnnotationReportFacetsByAccession", {
  # tests for AnnotationReportFacetsByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get genome annotation report summary information
  # Get genome annotation report summary information by genome accession. The return facets can be used in subsequent queries.
  # @param accession character 
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @return [V2GenomeAnnotationTableSummaryReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("AnnotationReportFacetsByPost", {
  # tests for AnnotationReportFacetsByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get genome annotation report summary information
  # Get genome annotation report summary information by genome accession. The return facets can be used in subsequent queries.
  # @param v2_genome_annotation_request V2GenomeAnnotationRequest 
  # @return [V2GenomeAnnotationTableSummaryReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("AssemblyAccessionsForSequenceAccession", {
  # tests for AssemblyAccessionsForSequenceAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get assembly accessions for a sequence accession
  # Get assembly accessions for a sequence (nucleotide) accession
  # @param accession character 
  # @return [V2AssemblyAccessions]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("AssemblyAccessionsForSequenceAccessionByPost", {
  # tests for AssemblyAccessionsForSequenceAccessionByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get assembly accessions for a sequence accession
  # Get assembly accessions for a sequence (nucleotide) accession
  # @param v2_sequence_accession_request V2SequenceAccessionRequest 
  # @return [V2AssemblyAccessions]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("AssemblyRevisionHistoryByGet", {
  # tests for AssemblyRevisionHistoryByGet
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get revision history for assembly by accession
  # Get revision history for assembly by accession
  # @param accession character 
  # @return [V2AssemblyRevisionHistory]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("AssemblyRevisionHistoryByPost", {
  # tests for AssemblyRevisionHistoryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get revision history for assembly by accession
  # Get revision history for assembly by accession
  # @param v2_assembly_revision_history_request V2AssemblyRevisionHistoryRequest 
  # @return [V2AssemblyRevisionHistory]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("CheckAssemblyAvailability", {
  # tests for CheckAssemblyAvailability
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Check the validity of genome accessions
  # The &#39;GET&#39; version of check is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param accessions array[character] NCBI genome assembly accessions
  # @return [V2AssemblyDatasetAvailability]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("CheckAssemblyAvailabilityPost", {
  # tests for CheckAssemblyAvailabilityPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Check the validity of many genome accessions in a single request
  # The &#39;GET&#39; version of check is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param v2_assembly_dataset_request V2AssemblyDatasetRequest 
  # @return [V2AssemblyDatasetAvailability]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("CheckmHistogramByTaxon", {
  # tests for CheckmHistogramByTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get CheckM histogram by species taxon
  # Get CheckM histogram by species taxon. CheckM histograms are only available for certain bacterial species.
  # @param species_taxon character 
  # @return [V2AssemblyCheckMHistogramReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("CheckmHistogramByTaxonByPost", {
  # tests for CheckmHistogramByTaxonByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get CheckM histogram by species taxon
  # Get CheckM histogram by species taxon. CheckM histograms are only available for certain bacterial species.
  # @param v2_assembly_check_m_histogram_request V2AssemblyCheckMHistogramRequest 
  # @return [V2AssemblyCheckMHistogramReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadAssemblyPackage", {
  # tests for DownloadAssemblyPackage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a genome dataset by accession
  # Download a genome dataset including fasta sequence, annotation and a detailed data report by accession.
  # @param accessions array[character] NCBI genome assembly accessions
  # @param chromosomes array[character] The default setting is all chromosome. Specify individual chromosome by string (1,2,MT or chr1,chr2.chrMT). Unplaced sequences are treated like their own chromosome ('Un'). The filter only applies to fasta sequence. (optional)
  # @param include_annotation_type array[V2AnnotationForAssemblyType] Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (optional)
  # @param hydrated V2AssemblyDatasetRequestResolution Set to DATA_REPORT_ONLY, to only retrieve data-reports. (optional)
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadAssemblyPackagePost", {
  # tests for DownloadAssemblyPackagePost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a genome dataset by post
  # The &#39;GET&#39; version of download is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param v2_assembly_dataset_request V2AssemblyDatasetRequest 
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadGenomeAnnotationPackage", {
  # tests for DownloadGenomeAnnotationPackage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get an annotation report dataset by accession
  # Download an annotation report including fasta sequence and a detailed annotation report by genomic accession.
  # @param accession character 
  # @param annotation_ids array[character] Limit the reports by internal, unstable annotation ids. (optional)
  # @param symbols array[character] Filter parameters (optional)
  # @param locations array[character] Locations with a chromosome or accession and optional start-stop range: chromosome|accession[:start-end] (optional)
  # @param gene_types array[character] granular gene_types (optional)
  # @param search_text array[character]  (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_annotation_type array[V2GenomeAnnotationRequestAnnotationType]  (optional)
  # @param page_size integer The maximum number of features to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format V2GenomeAnnotationRequestGenomeAnnotationTableFormat Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_token character A page token is returned from a `GetFeatures` call with more than `page_size` results. Use this token, along with the previous `FeatureRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadGenomeAnnotationPackageByPost", {
  # tests for DownloadGenomeAnnotationPackageByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get an annotation report dataset by accession
  # Download an annotation report including fasta sequence and a detailed annotation report by genomic accession.
  # @param v2_genome_annotation_request V2GenomeAnnotationRequest 
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeAnnotationDownloadSummary", {
  # tests for GenomeAnnotationDownloadSummary
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Preview feature dataset download
  # Get a download feature summary by accession in a JSON output format.
  # @param accession character 
  # @param annotation_ids array[character] Limit the reports by internal, unstable annotation ids. (optional)
  # @param symbols array[character] Filter parameters (optional)
  # @param locations array[character] Locations with a chromosome or accession and optional start-stop range: chromosome|accession[:start-end] (optional)
  # @param gene_types array[character] granular gene_types (optional)
  # @param search_text array[character]  (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_annotation_type array[V2GenomeAnnotationRequestAnnotationType]  (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format V2GenomeAnnotationRequestGenomeAnnotationTableFormat Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeAnnotationDownloadSummaryByPost", {
  # tests for GenomeAnnotationDownloadSummaryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Preview feature download by POST
  # The &#39;GET&#39; version of feature download summary is limited by the size of the GET URL (2KB).  The POST operation is provided to allow users to supply a larger number of annotation_ids in a single request.
  # @param v2_genome_annotation_request V2GenomeAnnotationRequest 
  # @return [V2DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeAnnotationReport", {
  # tests for GenomeAnnotationReport
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get genome annotation reports by genome accession
  # Get genome annotation reports by genome accession.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param accession character 
  # @param annotation_ids array[character] Limit the reports by internal, unstable annotation ids. (optional)
  # @param symbols array[character] Filter parameters (optional)
  # @param locations array[character] Locations with a chromosome or accession and optional start-stop range: chromosome|accession[:start-end] (optional)
  # @param gene_types array[character] granular gene_types (optional)
  # @param search_text array[character]  (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param page_size integer The maximum number of features to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param table_fields array[character] Specify which fields to include in the tabular report (optional)
  # @param table_format V2GenomeAnnotationRequestGenomeAnnotationTableFormat Optional pre-defined template for processing a tabular data request (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_token character A page token is returned from a `GetFeatures` call with more than `page_size` results. Use this token, along with the previous `FeatureRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @return [V2reportsGenomeAnnotationReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeAnnotationReportByPost", {
  # tests for GenomeAnnotationReportByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get genome annotation reports by genome accession
  # Get genome annotation reports by genome accession.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param v2_genome_annotation_request V2GenomeAnnotationRequest 
  # @return [V2reportsGenomeAnnotationReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReport", {
  # tests for GenomeDatasetReport
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by accessions
  # Get dataset reports by accessions.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param accessions array[character] 
  # @param filters_reference_only character If true, only return reference genome assemblies (optional)
  # @param filters_assembly_source V2AssemblyDatasetDescriptorsFilterAssemblySource Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies (optional)
  # @param filters_has_annotation character Return only annotated genome assemblies (optional)
  # @param filters_exclude_paired_reports character For paired (GCA/GCF) records, only return the primary record (optional)
  # @param filters_exclude_atypical character If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical (optional)
  # @param filters_assembly_version V2AssemblyDatasetDescriptorsFilterAssemblyVersion Return all assemblies, including replaced and suppressed, or only current assemblies (optional)
  # @param filters_assembly_level array[V2reportsAssemblyLevel] Only return genome assemblies that have one of the specified assembly levels. By default, do not filter. (optional)
  # @param filters_first_release_date character Only return genome assemblies that were released on or after the specified date By default, do not filter. (optional)
  # @param filters_last_release_date character Only return genome assemblies that were released on or before to the specified date By default, do not filter. (optional)
  # @param filters_search_text array[character] Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter (optional)
  # @param filters_is_metagenome_derived V2AssemblyDatasetDescriptorsFilterMetagenomeDerivedFilter  (optional)
  # @param filters_is_type_material character If true, include only type materials (optional)
  # @param filters_is_ictv_exemplar character If true, include only ICTV Exemplars (optional)
  # @param filters_exclude_multi_isolate character If true, exclude large multi-isolate projects (optional)
  # @param filters_type_material_category V2AssemblyDatasetDescriptorsFilterTypeMaterialCategory  (optional)
  # @param tax_exact_match character If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. (optional)
  # @param table_fields array[character]  (optional)
  # @param returned_content V2AssemblyDatasetReportsRequestContentType Return either assembly accessions, or complete assembly reports (optional)
  # @param page_size integer The maximum number of genome assembly reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReportByPost", {
  # tests for GenomeDatasetReportByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by accessions
  # Get a dataset report by accession.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param v2_assembly_dataset_reports_request V2AssemblyDatasetReportsRequest 
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReportsByAssemblyName", {
  # tests for GenomeDatasetReportsByAssemblyName
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by assembly name (exact)
  # Get dataset reports by assembly name (exact).  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param assembly_names array[character] 
  # @param filters_reference_only character If true, only return reference genome assemblies (optional)
  # @param filters_assembly_source V2AssemblyDatasetDescriptorsFilterAssemblySource Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies (optional)
  # @param filters_has_annotation character Return only annotated genome assemblies (optional)
  # @param filters_exclude_paired_reports character For paired (GCA/GCF) records, only return the primary record (optional)
  # @param filters_exclude_atypical character If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical (optional)
  # @param filters_assembly_version V2AssemblyDatasetDescriptorsFilterAssemblyVersion Return all assemblies, including replaced and suppressed, or only current assemblies (optional)
  # @param filters_assembly_level array[V2reportsAssemblyLevel] Only return genome assemblies that have one of the specified assembly levels. By default, do not filter. (optional)
  # @param filters_first_release_date character Only return genome assemblies that were released on or after the specified date By default, do not filter. (optional)
  # @param filters_last_release_date character Only return genome assemblies that were released on or before to the specified date By default, do not filter. (optional)
  # @param filters_search_text array[character] Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter (optional)
  # @param filters_is_metagenome_derived V2AssemblyDatasetDescriptorsFilterMetagenomeDerivedFilter  (optional)
  # @param filters_is_type_material character If true, include only type materials (optional)
  # @param filters_is_ictv_exemplar character If true, include only ICTV Exemplars (optional)
  # @param filters_exclude_multi_isolate character If true, exclude large multi-isolate projects (optional)
  # @param filters_type_material_category V2AssemblyDatasetDescriptorsFilterTypeMaterialCategory  (optional)
  # @param tax_exact_match character If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. (optional)
  # @param table_fields array[character]  (optional)
  # @param returned_content V2AssemblyDatasetReportsRequestContentType Return either assembly accessions, or complete assembly reports (optional)
  # @param page_size integer The maximum number of genome assembly reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReportsByBioproject", {
  # tests for GenomeDatasetReportsByBioproject
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by bioproject
  # Get dataset reports by bioprojects.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param bioprojects array[character] 
  # @param filters_reference_only character If true, only return reference genome assemblies (optional)
  # @param filters_assembly_source V2AssemblyDatasetDescriptorsFilterAssemblySource Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies (optional)
  # @param filters_has_annotation character Return only annotated genome assemblies (optional)
  # @param filters_exclude_paired_reports character For paired (GCA/GCF) records, only return the primary record (optional)
  # @param filters_exclude_atypical character If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical (optional)
  # @param filters_assembly_version V2AssemblyDatasetDescriptorsFilterAssemblyVersion Return all assemblies, including replaced and suppressed, or only current assemblies (optional)
  # @param filters_assembly_level array[V2reportsAssemblyLevel] Only return genome assemblies that have one of the specified assembly levels. By default, do not filter. (optional)
  # @param filters_first_release_date character Only return genome assemblies that were released on or after the specified date By default, do not filter. (optional)
  # @param filters_last_release_date character Only return genome assemblies that were released on or before to the specified date By default, do not filter. (optional)
  # @param filters_search_text array[character] Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter (optional)
  # @param filters_is_metagenome_derived V2AssemblyDatasetDescriptorsFilterMetagenomeDerivedFilter  (optional)
  # @param filters_is_type_material character If true, include only type materials (optional)
  # @param filters_is_ictv_exemplar character If true, include only ICTV Exemplars (optional)
  # @param filters_exclude_multi_isolate character If true, exclude large multi-isolate projects (optional)
  # @param filters_type_material_category V2AssemblyDatasetDescriptorsFilterTypeMaterialCategory  (optional)
  # @param tax_exact_match character If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. (optional)
  # @param table_fields array[character]  (optional)
  # @param returned_content V2AssemblyDatasetReportsRequestContentType Return either assembly accessions, or complete assembly reports (optional)
  # @param page_size integer The maximum number of genome assembly reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReportsByBiosampleId", {
  # tests for GenomeDatasetReportsByBiosampleId
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by biosample id
  # Get dataset reports by biosample id.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param biosample_ids array[character] 
  # @param filters_reference_only character If true, only return reference genome assemblies (optional)
  # @param filters_assembly_source V2AssemblyDatasetDescriptorsFilterAssemblySource Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies (optional)
  # @param filters_has_annotation character Return only annotated genome assemblies (optional)
  # @param filters_exclude_paired_reports character For paired (GCA/GCF) records, only return the primary record (optional)
  # @param filters_exclude_atypical character If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical (optional)
  # @param filters_assembly_version V2AssemblyDatasetDescriptorsFilterAssemblyVersion Return all assemblies, including replaced and suppressed, or only current assemblies (optional)
  # @param filters_assembly_level array[V2reportsAssemblyLevel] Only return genome assemblies that have one of the specified assembly levels. By default, do not filter. (optional)
  # @param filters_first_release_date character Only return genome assemblies that were released on or after the specified date By default, do not filter. (optional)
  # @param filters_last_release_date character Only return genome assemblies that were released on or before to the specified date By default, do not filter. (optional)
  # @param filters_search_text array[character] Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter (optional)
  # @param filters_is_metagenome_derived V2AssemblyDatasetDescriptorsFilterMetagenomeDerivedFilter  (optional)
  # @param filters_is_type_material character If true, include only type materials (optional)
  # @param filters_is_ictv_exemplar character If true, include only ICTV Exemplars (optional)
  # @param filters_exclude_multi_isolate character If true, exclude large multi-isolate projects (optional)
  # @param filters_type_material_category V2AssemblyDatasetDescriptorsFilterTypeMaterialCategory  (optional)
  # @param tax_exact_match character If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. (optional)
  # @param table_fields array[character]  (optional)
  # @param returned_content V2AssemblyDatasetReportsRequestContentType Return either assembly accessions, or complete assembly reports (optional)
  # @param page_size integer The maximum number of genome assembly reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReportsByTaxon", {
  # tests for GenomeDatasetReportsByTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by taxons
  # Get dataset reports by taxons.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param taxons array[character] NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
  # @param filters_reference_only character If true, only return reference genome assemblies (optional)
  # @param filters_assembly_source V2AssemblyDatasetDescriptorsFilterAssemblySource Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies (optional)
  # @param filters_has_annotation character Return only annotated genome assemblies (optional)
  # @param filters_exclude_paired_reports character For paired (GCA/GCF) records, only return the primary record (optional)
  # @param filters_exclude_atypical character If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical (optional)
  # @param filters_assembly_version V2AssemblyDatasetDescriptorsFilterAssemblyVersion Return all assemblies, including replaced and suppressed, or only current assemblies (optional)
  # @param filters_assembly_level array[V2reportsAssemblyLevel] Only return genome assemblies that have one of the specified assembly levels. By default, do not filter. (optional)
  # @param filters_first_release_date character Only return genome assemblies that were released on or after the specified date By default, do not filter. (optional)
  # @param filters_last_release_date character Only return genome assemblies that were released on or before to the specified date By default, do not filter. (optional)
  # @param filters_search_text array[character] Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter (optional)
  # @param filters_is_metagenome_derived V2AssemblyDatasetDescriptorsFilterMetagenomeDerivedFilter  (optional)
  # @param filters_is_type_material character If true, include only type materials (optional)
  # @param filters_is_ictv_exemplar character If true, include only ICTV Exemplars (optional)
  # @param filters_exclude_multi_isolate character If true, exclude large multi-isolate projects (optional)
  # @param filters_type_material_category V2AssemblyDatasetDescriptorsFilterTypeMaterialCategory  (optional)
  # @param tax_exact_match character If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. (optional)
  # @param table_fields array[character]  (optional)
  # @param returned_content V2AssemblyDatasetReportsRequestContentType Return either assembly accessions, or complete assembly reports (optional)
  # @param page_size integer The maximum number of genome assembly reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDatasetReportsByWgs", {
  # tests for GenomeDatasetReportsByWgs
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get dataset reports by wgs accession
  # Get dataset reports by wgs accession.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param wgs_accessions array[character] 
  # @param filters_reference_only character If true, only return reference genome assemblies (optional)
  # @param filters_assembly_source V2AssemblyDatasetDescriptorsFilterAssemblySource Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies (optional)
  # @param filters_has_annotation character Return only annotated genome assemblies (optional)
  # @param filters_exclude_paired_reports character For paired (GCA/GCF) records, only return the primary record (optional)
  # @param filters_exclude_atypical character If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical (optional)
  # @param filters_assembly_version V2AssemblyDatasetDescriptorsFilterAssemblyVersion Return all assemblies, including replaced and suppressed, or only current assemblies (optional)
  # @param filters_assembly_level array[V2reportsAssemblyLevel] Only return genome assemblies that have one of the specified assembly levels. By default, do not filter. (optional)
  # @param filters_first_release_date character Only return genome assemblies that were released on or after the specified date By default, do not filter. (optional)
  # @param filters_last_release_date character Only return genome assemblies that were released on or before to the specified date By default, do not filter. (optional)
  # @param filters_search_text array[character] Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter (optional)
  # @param filters_is_metagenome_derived V2AssemblyDatasetDescriptorsFilterMetagenomeDerivedFilter  (optional)
  # @param filters_is_type_material character If true, include only type materials (optional)
  # @param filters_is_ictv_exemplar character If true, include only ICTV Exemplars (optional)
  # @param filters_exclude_multi_isolate character If true, exclude large multi-isolate projects (optional)
  # @param filters_type_material_category V2AssemblyDatasetDescriptorsFilterTypeMaterialCategory  (optional)
  # @param tax_exact_match character If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. (optional)
  # @param table_fields array[character]  (optional)
  # @param returned_content V2AssemblyDatasetReportsRequestContentType Return either assembly accessions, or complete assembly reports (optional)
  # @param page_size integer The maximum number of genome assembly reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param sort_field character  (optional)
  # @param sort_direction V2SortDirection  (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @return [V2reportsAssemblyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDownloadSummary", {
  # tests for GenomeDownloadSummary
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Preview genome dataset download
  # Get a download summary by accession in a JSON output format.
  # @param accessions array[character] NCBI genome assembly accessions
  # @param chromosomes array[character] The default setting is all chromosome. Specify individual chromosome by string (1,2,MT or chr1,chr2.chrMT). Unplaced sequences are treated like their own chromosome ('Un'). The filter only applies to fasta sequence. (optional)
  # @param include_annotation_type array[V2AnnotationForAssemblyType] Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (optional)
  # @return [V2DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDownloadSummaryByPost", {
  # tests for GenomeDownloadSummaryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Preview genome dataset download by POST
  # The &#39;GET&#39; version of download summary is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param v2_assembly_dataset_request V2AssemblyDatasetRequest 
  # @return [V2DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeLinksByAccession", {
  # tests for GenomeLinksByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get assembly links by accessions
  # Get links to available assembly resources by accessions.
  # @param accessions array[character] NCBI genome assembly accessions, limited to 1000
  # @return [V2AssemblyLinksReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeLinksByAccessionByPost", {
  # tests for GenomeLinksByAccessionByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get assembly links by accessions
  # Get links to available assembly resources by accessions.
  # @param v2_assembly_links_request V2AssemblyLinksRequest 
  # @return [V2AssemblyLinksReply]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeSequenceReport", {
  # tests for GenomeSequenceReport
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get sequence reports by accessions
  # Get a sequence report by accession.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param accession character 
  # @param chromosomes array[character]  (optional)
  # @param role_filters array[character]  (optional)
  # @param table_fields array[character]  (optional)
  # @param count_assembly_unplaced character  (optional)
  # @param page_size integer The maximum number of genome assemblies to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from an `GetSequenceReports` call with more than `page_size` results. Use this token, along with the previous `AssemblyMetadataRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param include_tabular_header V2IncludeTabularHeader  (optional)
  # @return [V2SequenceReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeSequenceReportByPost", {
  # tests for GenomeSequenceReportByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get sequence reports by accessions
  # Get a sequence report by accession.  By default, in paged JSON format, but also available as tabular (accept: text/tab-separated-values) or json-lines (accept: application/x-ndjson)
  # @param v2_assembly_sequence_reports_request V2AssemblySequenceReportsRequest 
  # @return [V2SequenceReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})
