# Automatically generated by openapi-generator (https://openapi-generator.tech)
# Please update as you see appropriate

context("Test TaxonomyApi")

api_instance <- TaxonomyApi$new()

test_that("DownloadTaxonomyPackage", {
  # tests for DownloadTaxonomyPackage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a taxonomy data package by tax ID
  # Download a taxonomy report and names data package.
  # @param tax_ids array[integer] 
  # @param aux_reports array[V2TaxonomyDatasetRequestTaxonomyReportType] list additional reports to include with download. TAXONOMY_REPORT is included by default. (optional)
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadTaxonomyPackageByPost", {
  # tests for DownloadTaxonomyPackageByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a taxonomy data package by tax_id
  # Download a taxonomy report and names data package.
  # @param v2_taxonomy_dataset_request V2TaxonomyDatasetRequest 
  # @param filename character Output file name. (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxNameQuery", {
  # tests for TaxNameQuery
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a list of taxonomy names and IDs given a partial taxonomic name
  # This endpoint retrieves a list of taxonomy names and IDs given a possibly partial taxonomic name of any rank.
  # @param taxon_query character NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
  # @param tax_rank_filter V2OrganismQueryRequestTaxRankFilter Set the scope of searched tax ranks when filtering by gene or genome.  Not used for 'all' (optional)
  # @param taxon_resource_filter V2OrganismQueryRequestTaxonResourceFilter Limit results to those with gene or genome counts (no filter by default) (optional)
  # @param exact_match character If true, only return results that exactly match the provided name or tax-id (optional)
  # @return [V2SciNameAndIds]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxNameQueryByPost", {
  # tests for TaxNameQueryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Get a list of taxonomy names and IDs given a partial taxonomic name
  # This endpoint retrieves a list of taxonomy names and IDs given a possibly partial taxonomic name of any rank, by post.
  # @param v2_organism_query_request V2OrganismQueryRequest 
  # @return [V2SciNameAndIds]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyDataReport", {
  # tests for TaxonomyDataReport
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get taxonomic data report
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get metadata about a taxonomic node including taxonomic identifiers, lineage information, child nodes, and gene and genome counts in JSON format.
  # @param taxons array[character] 
  # @param returned_content V2TaxonomyMetadataRequestContentType Return either tax-ids alone, or entire taxononmy-metadata records (optional)
  # @param page_size integer The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_token character A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param table_format V2TaxonomyMetadataRequestTableFormat  (optional)
  # @param children character Flag for tax explosion. (optional)
  # @param ranks array[V2reportsRankType] Only include taxons of the provided ranks. If empty, return all ranks. (optional)
  # @return [V2reportsTaxonomyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyDataReportPost", {
  # tests for TaxonomyDataReportPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get taxonomic names data report by post
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get metadata about a taxonomic node including taxonomic identifiers, lineage information, child nodes, and gene and genome counts in JSON format.
  # @param v2_taxonomy_metadata_request V2TaxonomyMetadataRequest 
  # @return [V2reportsTaxonomyDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyFilteredSubtree", {
  # tests for TaxonomyFilteredSubtree
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get a filtered taxonomic subtree
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get a filtered taxonomic subtree that includes the full parent lineage and all immediate children from the selected taxonomic ranks in JSON format.
  # @param taxons array[character] 
  # @param specified_limit character Limit to specified species (optional)
  # @param rank_limits array[V2reportsRankType] Limit to the provided ranks.  If empty, accept any rank. (optional)
  # @return [V2TaxonomyFilteredSubtreeResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyFilteredSubtreePost", {
  # tests for TaxonomyFilteredSubtreePost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get a filtered taxonomic subtree by post
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get a filtered taxonomic subtree that includes the full parent lineage and all immediate children from the selected taxonomic ranks in JSON format.
  # @param v2_taxonomy_filtered_subtree_request V2TaxonomyFilteredSubtreeRequest 
  # @return [V2TaxonomyFilteredSubtreeResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyImage", {
  # tests for TaxonomyImage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Retrieve image associated with a taxonomic identifier
  # Using an NCBI Taxonomy ID or a name (common or scientific) at any rank, get the image associated with the taxon.
  # @param taxon character 
  # @param image_size V2ImageSize  (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyImageMetadata", {
  # tests for TaxonomyImageMetadata
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Retrieve image metadata associated with a taxonomic identifier
  # Using an NCBI Taxonomy ID or a name (common or scientific) at any rank, get the image metadata associated with the taxon.
  # @param taxon character 
  # @return [V2TaxonomyImageMetadataResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyImageMetadataPost", {
  # tests for TaxonomyImageMetadataPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Retrieve image metadata associated with a taxonomic identifier by post
  # Using an NCBI Taxonomy ID or a name (common or scientific) at any rank, get the image metadata associated with the taxon.
  # @param v2_taxonomy_image_metadata_request V2TaxonomyImageMetadataRequest 
  # @return [V2TaxonomyImageMetadataResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyImagePost", {
  # tests for TaxonomyImagePost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Retrieve image associated with a taxonomic identifier by post
  # Using an NCBI Taxonomy ID or a name (common or scientific) at any rank, get the image associated with the taxon.
  # @param v2_taxonomy_image_request V2TaxonomyImageRequest 
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyLinks", {
  # tests for TaxonomyLinks
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Retrieve external links associated with a taxonomic identifier.
  # Using an NCBI Taxonomy ID at any rank, get the external links associated with the taxon.
  # @param taxon character 
  # @return [V2TaxonomyLinksResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyLinksByPost", {
  # tests for TaxonomyLinksByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Retrieve external links associated with a taxonomic identifier.
  # Using an NCBI Taxonomy ID at any rank, get the external links associated with the taxon.
  # @param v2_taxonomy_links_request V2TaxonomyLinksRequest 
  # @return [V2TaxonomyLinksResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyMetadata", {
  # tests for TaxonomyMetadata
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get taxonomic metadata
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get metadata about a taxonomic node including taxonomic identifiers, lineage information, child nodes, and gene and genome counts in JSON format.
  # @param taxons array[character] 
  # @param returned_content V2TaxonomyMetadataRequestContentType Return either tax-ids alone, or entire taxononmy-metadata records (optional)
  # @param page_size integer The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_token character A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param table_format V2TaxonomyMetadataRequestTableFormat  (optional)
  # @param children character Flag for tax explosion. (optional)
  # @param ranks array[V2reportsRankType] Only include taxons of the provided ranks. If empty, return all ranks. (optional)
  # @return [V2TaxonomyMetadataResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyMetadataPost", {
  # tests for TaxonomyMetadataPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get taxonomic metadata by post
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get metadata about a taxonomic node including taxonomic identifiers, lineage information, child nodes, and gene and genome counts in JSON format.
  # @param v2_taxonomy_metadata_request V2TaxonomyMetadataRequest 
  # @return [V2TaxonomyMetadataResponse]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyNames", {
  # tests for TaxonomyNames
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get taxonomic names data report
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get metadata about associated taxonomic names.
  # @param taxons array[character] 
  # @param returned_content V2TaxonomyMetadataRequestContentType Return either tax-ids alone, or entire taxononmy-metadata records (optional)
  # @param page_size integer The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param include_tabular_header V2IncludeTabularHeader Whether this request for tabular data should include the header row (optional)
  # @param page_token character A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved. (optional)
  # @param table_format V2TaxonomyMetadataRequestTableFormat  (optional)
  # @param children character Flag for tax explosion. (optional)
  # @param ranks array[V2reportsRankType] Only include taxons of the provided ranks. If empty, return all ranks. (optional)
  # @return [V2reportsTaxonomyNamesDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyNamesPost", {
  # tests for TaxonomyNamesPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifiers to get taxonomic names data report by post
  # Using NCBI Taxonomy IDs or names (common or scientific) at any rank, get metadata about associated taxonomic names.
  # @param v2_taxonomy_metadata_request V2TaxonomyMetadataRequest 
  # @return [V2reportsTaxonomyNamesDataReportPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyRelatedIds", {
  # tests for TaxonomyRelatedIds
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifier to get related taxonomic identifiers, such as children
  # Using a single NCBI Taxonomy ID at any rank, get a list of related taxonomic IDs in JSON format.
  # @param tax_id integer 
  # @param include_lineage character If true, return reports for all taxonomy nodes in the lineages of the requested tax_id (optional)
  # @param include_subtree character This field is deprecated because all requests include the subtree, so it has no effect (optional)
  # @param ranks array[V2reportsRankType] Only include taxons of the provided ranks. If empty, return all ranks. (optional)
  # @param page_size integer The maximum number of taxids to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (optional)
  # @param page_token character A page token is returned from a `GetRelatedTaxids` call with more than `page_size` results. Use this token, along with the previous `TaxonomyRelatedIdRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (optional)
  # @return [V2TaxonomyTaxIdsPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("TaxonomyRelatedIdsPost", {
  # tests for TaxonomyRelatedIdsPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v2
  # Use taxonomic identifier to get related taxonomic identifiers, such as children
  # Using a single NCBI Taxonomy ID at any rank, get a list of related taxonomic IDs in JSON format.
  # @param v2_taxonomy_related_id_request V2TaxonomyRelatedIdRequest 
  # @return [V2TaxonomyTaxIdsPage]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})
