#' NCBI Datasets API
#'
#' ### NCBI Datasets is a resource that lets you easily gather data from NCBI. The Datasets version 2 API is still in alpha, and we're updating it often to add new functionality, iron out bugs and enhance usability. For some larger downloads, you may want to download a [dehydrated zip archive](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/how-tos/genomes/large-download/), and retrieve the individual data files at a later time. 
#'
#' The version of the OpenAPI document: v2
#' Generated by: https://openapi-generator.tech
#'
#' @docType class
#' @title Gene operations
#' @description GeneApi
#' @format An \code{R6Class} generator object
#' @field api_client Handles the client-server communication.
#'
#' @examples
#' \dontrun{
#' ####################  DownloadGenePackage  ####################
#'
#' library(ncbiopenapi)
#' var_gene_ids <- c(123) # array[integer] | NCBI gene ids
#' var_include_annotation_type <- c(v2Fasta$new()) # array[V2Fasta] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (Optional)
#' var_fasta_filter <- c("inner_example") # array[character] | Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated) (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions (Optional)
#' var_aux_report <- c(v2GeneDatasetRequestGeneDatasetReportType$new()) # array[V2GeneDatasetRequestGeneDatasetReportType] | list additional reports to include with download. Data report is included by default. (Optional)
#' var_tabular_reports <- c(v2GeneDatasetRequestGeneDatasetReportType$new()) # array[V2GeneDatasetRequestGeneDatasetReportType] |  (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead. (Optional)
#' var_table_report_type <- v2GeneDatasetRequestGeneDatasetReportType$new() # V2GeneDatasetRequestGeneDatasetReportType | Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead. (Optional)
#' var_filename <- "ncbi_dataset.zip" # character | Output file name. (Optional)
#'
#' #Get a gene dataset by gene ID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$DownloadGenePackage(var_gene_ids, include_annotation_type = var_include_annotation_type, fasta_filter = var_fasta_filter, accession_filter = var_accession_filter, aux_report = var_aux_report, tabular_reports = var_tabular_reports, table_fields = var_table_fields, table_report_type = var_table_report_type, filename = var_filenamedata_file = "result.txt")
#' result <- api_instance$DownloadGenePackage(var_gene_ids, include_annotation_type = var_include_annotation_type, fasta_filter = var_fasta_filter, accession_filter = var_accession_filter, aux_report = var_aux_report, tabular_reports = var_tabular_reports, table_fields = var_table_fields, table_report_type = var_table_report_type, filename = var_filename)
#' dput(result)
#'
#'
#' ####################  DownloadGenePackagePost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_dataset_request <- v2GeneDatasetRequest$new(c(123), c(v2Fasta$new()), v2GeneDatasetRequestContentType$new(), c("fasta_filter_example"), c("accession_filter_example"), c(v2GeneDatasetRequestGeneDatasetReportType$new()), c(v2GeneDatasetRequestGeneDatasetReportType$new()), c("table_fields_example"), v2GeneDatasetRequestGeneDatasetReportType$new()) # V2GeneDatasetRequest | 
#' var_filename <- "ncbi_dataset.zip" # character | Output file name. (Optional)
#'
#' #Get a gene dataset by POST
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$DownloadGenePackagePost(var_v2_gene_dataset_request, filename = var_filenamedata_file = "result.txt")
#' result <- api_instance$DownloadGenePackagePost(var_v2_gene_dataset_request, filename = var_filename)
#' dput(result)
#'
#'
#' ####################  GeneChromosomeSummary  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9117" # character | 
#' var_annotation_name <- "GCF_028858705.1-RS_2023_03" # character | 
#'
#' #Get summary of chromosomes for a particular taxon's annotation
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneChromosomeSummary(var_taxon, var_annotation_namedata_file = "result.txt")
#' result <- api_instance$GeneChromosomeSummary(var_taxon, var_annotation_name)
#' dput(result)
#'
#'
#' ####################  GeneCountsForTaxon  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | Taxon for provided gene symbol
#'
#' #Get gene counts by taxonomic identifier
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneCountsForTaxon(var_taxondata_file = "result.txt")
#' result <- api_instance$GeneCountsForTaxon(var_taxon)
#' dput(result)
#'
#'
#' ####################  GeneCountsForTaxonByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_counts_by_taxon_request <- v2GeneCountsByTaxonRequest$new("taxon_example") # V2GeneCountsByTaxonRequest | 
#'
#' #Get gene counts by taxonomic identifier
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneCountsForTaxonByPost(var_v2_gene_counts_by_taxon_requestdata_file = "result.txt")
#' result <- api_instance$GeneCountsForTaxonByPost(var_v2_gene_counts_by_taxon_request)
#' dput(result)
#'
#'
#' ####################  GeneDatasetReport  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_dataset_reports_request <- v2GeneDatasetReportsRequest$new(v2GeneDatasetReportsRequestContentType$new(), c(123), c("accessions_example"), v2GeneDatasetReportsRequestSymbolsForTaxon$new(c("symbols_example"), "taxon_example"), "taxon_example", c("locus_tags_example"), c("table_fields_example"), "table_format_example", v2IncludeTabularHeader$new(), 123, "page_token_example", "query_example", c(v2GeneType$new()), c("accession_filter_example"), "tax_search_subtree_example", c(v2SortField$new("field_example", v2SortDirection$new()))) # V2GeneDatasetReportsRequest | 
#'
#' #Get gene dataset report as JSON
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDatasetReport(var_v2_gene_dataset_reports_requestdata_file = "result.txt")
#' result <- api_instance$GeneDatasetReport(var_v2_gene_dataset_reports_request)
#' dput(result)
#'
#'
#' ####################  GeneDatasetReportByAccession  ####################
#'
#' library(ncbiopenapi)
#' var_accessions <- c("inner_example") # array[character] | RNA or Protein accessions.
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get dataset reports by accession IDs
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDatasetReportByAccession(var_accessions, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneDatasetReportByAccession(var_accessions, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneDatasetReportByTaxAndSymbol  ####################
#'
#' library(ncbiopenapi)
#' var_symbols <- c("inner_example") # array[character] | Gene symbol
#' var_taxon <- "9606" # character | Taxon for provided gene symbol
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get dataset reports by taxons.
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDatasetReportByTaxAndSymbol(var_symbols, var_taxon, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneDatasetReportByTaxAndSymbol(var_symbols, var_taxon, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneDatasetReportsById  ####################
#'
#' library(ncbiopenapi)
#' var_gene_ids <- c(123) # array[integer] | NCBI gene ids
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get dataset reports by gene IDs.
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDatasetReportsById(var_gene_ids, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneDatasetReportsById(var_gene_ids, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneDatasetReportsByLocusTag  ####################
#'
#' library(ncbiopenapi)
#' var_locus_tags <- c("inner_example") # array[character] | Gene locus tags
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene dataset reports by locus tag
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDatasetReportsByLocusTag(var_locus_tags, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneDatasetReportsByLocusTag(var_locus_tags, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneDatasetReportsByTaxon  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene dataset reports by taxonomic identifier
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDatasetReportsByTaxon(var_taxon, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneDatasetReportsByTaxon(var_taxon, returned_content = var_returned_content, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneDownloadSummaryById  ####################
#'
#' library(ncbiopenapi)
#' var_gene_ids <- c(123) # array[integer] | NCBI gene ids
#' var_include_annotation_type <- c(v2Fasta$new()) # array[V2Fasta] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (Optional)
#' var_returned_content <- v2GeneDatasetRequestContentType$new() # V2GeneDatasetRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_fasta_filter <- c("inner_example") # array[character] | Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated) (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions (Optional)
#' var_aux_report <- c(v2GeneDatasetRequestGeneDatasetReportType$new()) # array[V2GeneDatasetRequestGeneDatasetReportType] | list additional reports to include with download. Data report is included by default. (Optional)
#' var_tabular_reports <- c(v2GeneDatasetRequestGeneDatasetReportType$new()) # array[V2GeneDatasetRequestGeneDatasetReportType] |  (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead. (Optional)
#' var_table_report_type <- v2GeneDatasetRequestGeneDatasetReportType$new() # V2GeneDatasetRequestGeneDatasetReportType | Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead. (Optional)
#'
#' #Get gene download summary by GeneID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDownloadSummaryById(var_gene_ids, include_annotation_type = var_include_annotation_type, returned_content = var_returned_content, fasta_filter = var_fasta_filter, accession_filter = var_accession_filter, aux_report = var_aux_report, tabular_reports = var_tabular_reports, table_fields = var_table_fields, table_report_type = var_table_report_typedata_file = "result.txt")
#' result <- api_instance$GeneDownloadSummaryById(var_gene_ids, include_annotation_type = var_include_annotation_type, returned_content = var_returned_content, fasta_filter = var_fasta_filter, accession_filter = var_accession_filter, aux_report = var_aux_report, tabular_reports = var_tabular_reports, table_fields = var_table_fields, table_report_type = var_table_report_type)
#' dput(result)
#'
#'
#' ####################  GeneDownloadSummaryByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_dataset_request <- v2GeneDatasetRequest$new(c(123), c(v2Fasta$new()), v2GeneDatasetRequestContentType$new(), c("fasta_filter_example"), c("accession_filter_example"), c(v2GeneDatasetRequestGeneDatasetReportType$new()), c(v2GeneDatasetRequestGeneDatasetReportType$new()), c("table_fields_example"), v2GeneDatasetRequestGeneDatasetReportType$new()) # V2GeneDatasetRequest | 
#'
#' #Get gene download summary
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneDownloadSummaryByPost(var_v2_gene_dataset_requestdata_file = "result.txt")
#' result <- api_instance$GeneDownloadSummaryByPost(var_v2_gene_dataset_request)
#' dput(result)
#'
#'
#' ####################  GeneLinksById  ####################
#'
#' library(ncbiopenapi)
#' var_gene_ids <- c(123) # array[integer] | NCBI gene ids, limited to 1000 ids
#'
#' #Get gene links by gene ID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneLinksById(var_gene_idsdata_file = "result.txt")
#' result <- api_instance$GeneLinksById(var_gene_ids)
#' dput(result)
#'
#'
#' ####################  GeneLinksByIdByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_links_request <- v2GeneLinksRequest$new(c(123)) # V2GeneLinksRequest | 
#'
#' #Get gene links by gene ID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneLinksByIdByPost(var_v2_gene_links_requestdata_file = "result.txt")
#' result <- api_instance$GeneLinksByIdByPost(var_v2_gene_links_request)
#' dput(result)
#'
#'
#' ####################  GeneMetadataByAccession  ####################
#'
#' library(ncbiopenapi)
#' var_accessions <- c("inner_example") # array[character] | RNA or Protein accessions.
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_locus_tags <- c("inner_example") # array[character] | Gene locus tags (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene metadata by RefSeq Accession
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneMetadataByAccession(var_accessions, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneMetadataByAccession(var_accessions, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneMetadataByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_dataset_reports_request <- v2GeneDatasetReportsRequest$new(v2GeneDatasetReportsRequestContentType$new(), c(123), c("accessions_example"), v2GeneDatasetReportsRequestSymbolsForTaxon$new(c("symbols_example"), "taxon_example"), "taxon_example", c("locus_tags_example"), c("table_fields_example"), "table_format_example", v2IncludeTabularHeader$new(), 123, "page_token_example", "query_example", c(v2GeneType$new()), c("accession_filter_example"), "tax_search_subtree_example", c(v2SortField$new("field_example", v2SortDirection$new()))) # V2GeneDatasetReportsRequest | 
#'
#' #Get gene metadata as JSON
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneMetadataByPost(var_v2_gene_dataset_reports_requestdata_file = "result.txt")
#' result <- api_instance$GeneMetadataByPost(var_v2_gene_dataset_reports_request)
#' dput(result)
#'
#'
#' ####################  GeneMetadataByTaxAndSymbol  ####################
#'
#' library(ncbiopenapi)
#' var_symbols <- c("inner_example") # array[character] | Gene symbol
#' var_taxon <- "9606" # character | Taxon for provided gene symbol
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_locus_tags <- c("inner_example") # array[character] | Gene locus tags (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene metadata by gene symbol
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneMetadataByTaxAndSymbol(var_symbols, var_taxon, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneMetadataByTaxAndSymbol(var_symbols, var_taxon, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneOrthologsById  ####################
#'
#' library(ncbiopenapi)
#' var_gene_id <- 2778 # integer | 
#' var_returned_content <- v2OrthologRequestContentType$new() # V2OrthologRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_taxon_filter <- c("inner_example") # array[character] | Filter genes by taxa (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `OrthologRequest` call with more than `page_size` results. Use this token, along with the previous `OrthologRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#'
#' #Get gene orthologs by gene ID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneOrthologsById(var_gene_id, returned_content = var_returned_content, taxon_filter = var_taxon_filter, page_size = var_page_size, page_token = var_page_tokendata_file = "result.txt")
#' result <- api_instance$GeneOrthologsById(var_gene_id, returned_content = var_returned_content, taxon_filter = var_taxon_filter, page_size = var_page_size, page_token = var_page_token)
#' dput(result)
#'
#'
#' ####################  GeneOrthologsByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_ortholog_request <- v2OrthologRequest$new(123, v2OrthologRequestContentType$new(), c("taxon_filter_example"), 123, "page_token_example") # V2OrthologRequest | 
#'
#' #Get gene orthologs by gene ID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneOrthologsByPost(var_v2_ortholog_requestdata_file = "result.txt")
#' result <- api_instance$GeneOrthologsByPost(var_v2_ortholog_request)
#' dput(result)
#'
#'
#' ####################  GeneProductReport  ####################
#'
#' library(ncbiopenapi)
#' var_v2_gene_dataset_reports_request <- v2GeneDatasetReportsRequest$new(v2GeneDatasetReportsRequestContentType$new(), c(123), c("accessions_example"), v2GeneDatasetReportsRequestSymbolsForTaxon$new(c("symbols_example"), "taxon_example"), "taxon_example", c("locus_tags_example"), c("table_fields_example"), "table_format_example", v2IncludeTabularHeader$new(), 123, "page_token_example", "query_example", c(v2GeneType$new()), c("accession_filter_example"), "tax_search_subtree_example", c(v2SortField$new("field_example", v2SortDirection$new()))) # V2GeneDatasetReportsRequest | 
#'
#' #Get gene product reports as JSON
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneProductReport(var_v2_gene_dataset_reports_requestdata_file = "result.txt")
#' result <- api_instance$GeneProductReport(var_v2_gene_dataset_reports_request)
#' dput(result)
#'
#'
#' ####################  GeneProductReportByAccession  ####################
#'
#' library(ncbiopenapi)
#' var_accessions <- c("inner_example") # array[character] | RNA or Protein accessions.
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene product reports by accession IDs
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneProductReportByAccession(var_accessions, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneProductReportByAccession(var_accessions, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneProductReportByTaxAndSymbol  ####################
#'
#' library(ncbiopenapi)
#' var_symbols <- c("inner_example") # array[character] | Gene symbol
#' var_taxon <- "9606" # character | Taxon for provided gene symbol
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get product reports by taxon.
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneProductReportByTaxAndSymbol(var_symbols, var_taxon, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneProductReportByTaxAndSymbol(var_symbols, var_taxon, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneProductReportsById  ####################
#'
#' library(ncbiopenapi)
#' var_gene_ids <- c(123) # array[integer] | NCBI gene ids
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene product reports by gene IDs.
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneProductReportsById(var_gene_ids, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneProductReportsById(var_gene_ids, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneProductReportsByLocusTags  ####################
#'
#' library(ncbiopenapi)
#' var_locus_tags <- c("inner_example") # array[character] | Gene locus tags
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene product reports by locus tags
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneProductReportsByLocusTags(var_locus_tags, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneProductReportsByLocusTags(var_locus_tags, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneProductReportsByTaxon  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_table_format <- "table_format_example" # character | Optional pre-defined template for processing a tabular data request (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene product reports by taxonomic identifier
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneProductReportsByTaxon(var_taxon, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneProductReportsByTaxon(var_taxon, table_fields = var_table_fields, table_format = var_table_format, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneReportsById  ####################
#'
#' library(ncbiopenapi)
#' var_gene_ids <- c(123) # array[integer] | NCBI gene ids
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_locus_tags <- c("inner_example") # array[character] | Gene locus tags (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene reports by GeneID
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneReportsById(var_gene_ids, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneReportsById(var_gene_ids, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' ####################  GeneReportsByTaxon  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
#' var_returned_content <- v2GeneDatasetReportsRequestContentType$new() # V2GeneDatasetReportsRequestContentType | Return either gene-ids, or entire gene metadata (Optional)
#' var_locus_tags <- c("inner_example") # array[character] | Gene locus tags (Optional)
#' var_table_fields <- c("inner_example") # array[character] | Specify which fields to include in the tabular report (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_size <- 20 # integer | The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_query <- "A2M immunoglobulin" # character | text search within gene symbol, aliases, name, locus-tag and protein name (Optional)
#' var_types <- c(v2GeneType$new()) # array[V2GeneType] | Gene types to filter (Optional)
#' var_accession_filter <- c("inner_example") # array[character] | Filter tabular product report to only include the selected accessions (Optional)
#' var_tax_search_subtree <- FALSE # character | For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (Optional)
#' var_sort_field <- "sort_field_example" # character |  (Optional)
#' var_sort_direction <- v2SortDirection$new() # V2SortDirection |  (Optional)
#'
#' #Get gene reports by taxonomic identifier
#' api_instance <- GeneApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$GeneReportsByTaxon(var_taxon, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_directiondata_file = "result.txt")
#' result <- api_instance$GeneReportsByTaxon(var_taxon, returned_content = var_returned_content, locus_tags = var_locus_tags, table_fields = var_table_fields, include_tabular_header = var_include_tabular_header, page_size = var_page_size, page_token = var_page_token, query = var_query, types = var_types, accession_filter = var_accession_filter, tax_search_subtree = var_tax_search_subtree, sort_field = var_sort_field, sort_direction = var_sort_direction)
#' dput(result)
#'
#'
#' }
#' @importFrom R6 R6Class
#' @importFrom base64enc base64encode
#' @export
GeneApi <- R6::R6Class(
  "GeneApi",
  public = list(
    api_client = NULL,

    #' @description
    #' Initialize a new GeneApi.
    #'
    #' @param api_client An instance of API client.
    initialize = function(api_client) {
      if (!missing(api_client)) {
        self$api_client <- api_client
      } else {
        self$api_client <- ApiClient$new()
      }
    },

    #' @description
    #' Get a gene dataset by gene ID
    #'
    #' @param gene_ids NCBI gene ids
    #' @param include_annotation_type (optional) Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
    #' @param fasta_filter (optional) Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated)
    #' @param accession_filter (optional) Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions
    #' @param aux_report (optional) list additional reports to include with download. Data report is included by default.
    #' @param tabular_reports (optional) No description
    #' @param table_fields (optional) Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead.
    #' @param table_report_type (optional) Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead.
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return data.frame
    DownloadGenePackage = function(gene_ids, include_annotation_type = NULL, fasta_filter = NULL, accession_filter = NULL, aux_report = NULL, tabular_reports = NULL, table_fields = NULL, table_report_type = NULL, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      local_var_response <- self$DownloadGenePackageWithHttpInfo(gene_ids, include_annotation_type, fasta_filter, accession_filter, aux_report, tabular_reports, table_fields, table_report_type, filename, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get a gene dataset by gene ID
    #'
    #' @param gene_ids NCBI gene ids
    #' @param include_annotation_type (optional) Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
    #' @param fasta_filter (optional) Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated)
    #' @param accession_filter (optional) Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions
    #' @param aux_report (optional) list additional reports to include with download. Data report is included by default.
    #' @param tabular_reports (optional) No description
    #' @param table_fields (optional) Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead.
    #' @param table_report_type (optional) Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead.
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (data.frame) with additional information such as HTTP status code, headers
    DownloadGenePackageWithHttpInfo = function(gene_ids, include_annotation_type = NULL, fasta_filter = NULL, accession_filter = NULL, aux_report = NULL, tabular_reports = NULL, table_fields = NULL, table_report_type = NULL, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_ids`)) {
        stop("Missing required parameter `gene_ids`.")
      }










      # explore
      for (query_item in `include_annotation_type`) {
        query_params[["include_annotation_type"]] <- c(query_params[["include_annotation_type"]], list(`include_annotation_type` = query_item))
      }

      # explore
      for (query_item in `fasta_filter`) {
        query_params[["fasta_filter"]] <- c(query_params[["fasta_filter"]], list(`fasta_filter` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      # explore
      for (query_item in `aux_report`) {
        query_params[["aux_report"]] <- c(query_params[["aux_report"]], list(`aux_report` = query_item))
      }

      # explore
      for (query_item in `tabular_reports`) {
        query_params[["tabular_reports"]] <- c(query_params[["tabular_reports"]], list(`tabular_reports` = query_item))
      }

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_report_type"]] <- `table_report_type`

      query_params[["filename"]] <- `filename`

      local_var_url_path <- "/gene/id/{gene_ids}/download"
      if (!missing(`gene_ids`)) {
        local_var_url_path <- gsub("\\{gene_ids\\}", paste(URLencode(as.character(`gene_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/zip")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "data.frame"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get a gene dataset by POST
    #'
    #' @param v2_gene_dataset_request 
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return data.frame
    DownloadGenePackagePost = function(v2_gene_dataset_request, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      local_var_response <- self$DownloadGenePackagePostWithHttpInfo(v2_gene_dataset_request, filename, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get a gene dataset by POST
    #'
    #' @param v2_gene_dataset_request 
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (data.frame) with additional information such as HTTP status code, headers
    DownloadGenePackagePostWithHttpInfo = function(v2_gene_dataset_request, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_dataset_request`)) {
        stop("Missing required parameter `v2_gene_dataset_request`.")
      }



      query_params[["filename"]] <- `filename`

      if (!is.null(`v2_gene_dataset_request`)) {
        local_var_body <- `v2_gene_dataset_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/download"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/zip")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "data.frame"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get summary of chromosomes for a particular taxon's annotation
    #'
    #' @param taxon 
    #' @param annotation_name 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2GeneChromosomeSummaryReply
    GeneChromosomeSummary = function(taxon, annotation_name, data_file = NULL, ...) {
      local_var_response <- self$GeneChromosomeSummaryWithHttpInfo(taxon, annotation_name, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get summary of chromosomes for a particular taxon's annotation
    #'
    #' @param taxon 
    #' @param annotation_name 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2GeneChromosomeSummaryReply) with additional information such as HTTP status code, headers
    GeneChromosomeSummaryWithHttpInfo = function(taxon, annotation_name, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      if (missing(`annotation_name`)) {
        stop("Missing required parameter `annotation_name`.")
      }



      local_var_url_path <- "/gene/taxon/{taxon}/annotation/{annotation_name}/chromosome_summary"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      if (!missing(`annotation_name`)) {
        local_var_url_path <- gsub("\\{annotation_name\\}", URLencode(as.character(`annotation_name`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2GeneChromosomeSummaryReply"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene counts by taxonomic identifier
    #'
    #' @param taxon Taxon for provided gene symbol
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2GeneCountsByTaxonReply
    GeneCountsForTaxon = function(taxon, data_file = NULL, ...) {
      local_var_response <- self$GeneCountsForTaxonWithHttpInfo(taxon, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene counts by taxonomic identifier
    #'
    #' @param taxon Taxon for provided gene symbol
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2GeneCountsByTaxonReply) with additional information such as HTTP status code, headers
    GeneCountsForTaxonWithHttpInfo = function(taxon, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }


      local_var_url_path <- "/gene/taxon/{taxon}/counts"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2GeneCountsByTaxonReply"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene counts by taxonomic identifier
    #'
    #' @param v2_gene_counts_by_taxon_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2GeneCountsByTaxonReply
    GeneCountsForTaxonByPost = function(v2_gene_counts_by_taxon_request, data_file = NULL, ...) {
      local_var_response <- self$GeneCountsForTaxonByPostWithHttpInfo(v2_gene_counts_by_taxon_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene counts by taxonomic identifier
    #'
    #' @param v2_gene_counts_by_taxon_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2GeneCountsByTaxonReply) with additional information such as HTTP status code, headers
    GeneCountsForTaxonByPostWithHttpInfo = function(v2_gene_counts_by_taxon_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_counts_by_taxon_request`)) {
        stop("Missing required parameter `v2_gene_counts_by_taxon_request`.")
      }


      if (!is.null(`v2_gene_counts_by_taxon_request`)) {
        local_var_body <- `v2_gene_counts_by_taxon_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/taxon/counts"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2GeneCountsByTaxonReply"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene dataset report as JSON
    #'
    #' @param v2_gene_dataset_reports_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneDatasetReport = function(v2_gene_dataset_reports_request, data_file = NULL, ...) {
      local_var_response <- self$GeneDatasetReportWithHttpInfo(v2_gene_dataset_reports_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene dataset report as JSON
    #'
    #' @param v2_gene_dataset_reports_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneDatasetReportWithHttpInfo = function(v2_gene_dataset_reports_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_dataset_reports_request`)) {
        stop("Missing required parameter `v2_gene_dataset_reports_request`.")
      }


      if (!is.null(`v2_gene_dataset_reports_request`)) {
        local_var_body <- `v2_gene_dataset_reports_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/dataset_report"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get dataset reports by accession IDs
    #'
    #' @param accessions RNA or Protein accessions.
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneDatasetReportByAccession = function(accessions, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneDatasetReportByAccessionWithHttpInfo(accessions, returned_content, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get dataset reports by accession IDs
    #'
    #' @param accessions RNA or Protein accessions.
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneDatasetReportByAccessionWithHttpInfo = function(accessions, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`accessions`)) {
        stop("Missing required parameter `accessions`.")
      }













      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/accession/{accessions}/dataset_report"
      if (!missing(`accessions`)) {
        local_var_url_path <- gsub("\\{accessions\\}", paste(URLencode(as.character(`accessions`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get dataset reports by taxons.
    #'
    #' @param symbols Gene symbol
    #' @param taxon Taxon for provided gene symbol
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneDatasetReportByTaxAndSymbol = function(symbols, taxon, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneDatasetReportByTaxAndSymbolWithHttpInfo(symbols, taxon, returned_content, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get dataset reports by taxons.
    #'
    #' @param symbols Gene symbol
    #' @param taxon Taxon for provided gene symbol
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneDatasetReportByTaxAndSymbolWithHttpInfo = function(symbols, taxon, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`symbols`)) {
        stop("Missing required parameter `symbols`.")
      }

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }














      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/symbol/{symbols}/taxon/{taxon}/dataset_report"
      if (!missing(`symbols`)) {
        local_var_url_path <- gsub("\\{symbols\\}", paste(URLencode(as.character(`symbols`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get dataset reports by gene IDs.
    #'
    #' @param gene_ids NCBI gene ids
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneDatasetReportsById = function(gene_ids, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneDatasetReportsByIdWithHttpInfo(gene_ids, returned_content, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get dataset reports by gene IDs.
    #'
    #' @param gene_ids NCBI gene ids
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneDatasetReportsByIdWithHttpInfo = function(gene_ids, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_ids`)) {
        stop("Missing required parameter `gene_ids`.")
      }













      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/id/{gene_ids}/dataset_report"
      if (!missing(`gene_ids`)) {
        local_var_url_path <- gsub("\\{gene_ids\\}", paste(URLencode(as.character(`gene_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene dataset reports by locus tag
    #'
    #' @param locus_tags Gene locus tags
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneDatasetReportsByLocusTag = function(locus_tags, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneDatasetReportsByLocusTagWithHttpInfo(locus_tags, returned_content, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene dataset reports by locus tag
    #'
    #' @param locus_tags Gene locus tags
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneDatasetReportsByLocusTagWithHttpInfo = function(locus_tags, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`locus_tags`)) {
        stop("Missing required parameter `locus_tags`.")
      }













      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/locus_tag/{locus_tags}/dataset_report"
      if (!missing(`locus_tags`)) {
        local_var_url_path <- gsub("\\{locus_tags\\}", paste(URLencode(as.character(`locus_tags`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene dataset reports by taxonomic identifier
    #'
    #' @param taxon NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneDatasetReportsByTaxon = function(taxon, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneDatasetReportsByTaxonWithHttpInfo(taxon, returned_content, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene dataset reports by taxonomic identifier
    #'
    #' @param taxon NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneDatasetReportsByTaxonWithHttpInfo = function(taxon, returned_content = NULL, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }













      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/taxon/{taxon}/dataset_report"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene download summary by GeneID
    #'
    #' @param gene_ids NCBI gene ids
    #' @param include_annotation_type (optional) Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param fasta_filter (optional) Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated)
    #' @param accession_filter (optional) Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions
    #' @param aux_report (optional) list additional reports to include with download. Data report is included by default.
    #' @param tabular_reports (optional) No description
    #' @param table_fields (optional) Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead.
    #' @param table_report_type (optional) Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2DownloadSummary
    GeneDownloadSummaryById = function(gene_ids, include_annotation_type = NULL, returned_content = NULL, fasta_filter = NULL, accession_filter = NULL, aux_report = NULL, tabular_reports = NULL, table_fields = NULL, table_report_type = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneDownloadSummaryByIdWithHttpInfo(gene_ids, include_annotation_type, returned_content, fasta_filter, accession_filter, aux_report, tabular_reports, table_fields, table_report_type, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene download summary by GeneID
    #'
    #' @param gene_ids NCBI gene ids
    #' @param include_annotation_type (optional) Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param fasta_filter (optional) Limit the FASTA sequences in the datasets package to these transcript and protein accessions (deprecated)
    #' @param accession_filter (optional) Limit the FASTA sequences and tabular product report in the datasets package to these transcript and protein accessions
    #' @param aux_report (optional) list additional reports to include with download. Data report is included by default.
    #' @param tabular_reports (optional) No description
    #' @param table_fields (optional) Specify which fields to include in the tabular report. This is currently deprecated - use tabular_reports instead.
    #' @param table_report_type (optional) Specify the report from which the table fields will be taken. This is currently deprecated - use tabular_reports instead.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2DownloadSummary) with additional information such as HTTP status code, headers
    GeneDownloadSummaryByIdWithHttpInfo = function(gene_ids, include_annotation_type = NULL, returned_content = NULL, fasta_filter = NULL, accession_filter = NULL, aux_report = NULL, tabular_reports = NULL, table_fields = NULL, table_report_type = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_ids`)) {
        stop("Missing required parameter `gene_ids`.")
      }










      # explore
      for (query_item in `include_annotation_type`) {
        query_params[["include_annotation_type"]] <- c(query_params[["include_annotation_type"]], list(`include_annotation_type` = query_item))
      }

      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `fasta_filter`) {
        query_params[["fasta_filter"]] <- c(query_params[["fasta_filter"]], list(`fasta_filter` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      # explore
      for (query_item in `aux_report`) {
        query_params[["aux_report"]] <- c(query_params[["aux_report"]], list(`aux_report` = query_item))
      }

      # explore
      for (query_item in `tabular_reports`) {
        query_params[["tabular_reports"]] <- c(query_params[["tabular_reports"]], list(`tabular_reports` = query_item))
      }

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_report_type"]] <- `table_report_type`

      local_var_url_path <- "/gene/id/{gene_ids}/download_summary"
      if (!missing(`gene_ids`)) {
        local_var_url_path <- gsub("\\{gene_ids\\}", paste(URLencode(as.character(`gene_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2DownloadSummary"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene download summary
    #'
    #' @param v2_gene_dataset_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2DownloadSummary
    GeneDownloadSummaryByPost = function(v2_gene_dataset_request, data_file = NULL, ...) {
      local_var_response <- self$GeneDownloadSummaryByPostWithHttpInfo(v2_gene_dataset_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene download summary
    #'
    #' @param v2_gene_dataset_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2DownloadSummary) with additional information such as HTTP status code, headers
    GeneDownloadSummaryByPostWithHttpInfo = function(v2_gene_dataset_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_dataset_request`)) {
        stop("Missing required parameter `v2_gene_dataset_request`.")
      }


      if (!is.null(`v2_gene_dataset_request`)) {
        local_var_body <- `v2_gene_dataset_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/download_summary"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2DownloadSummary"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene links by gene ID
    #'
    #' @param gene_ids NCBI gene ids, limited to 1000 ids
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2GeneLinksReply
    GeneLinksById = function(gene_ids, data_file = NULL, ...) {
      local_var_response <- self$GeneLinksByIdWithHttpInfo(gene_ids, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene links by gene ID
    #'
    #' @param gene_ids NCBI gene ids, limited to 1000 ids
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2GeneLinksReply) with additional information such as HTTP status code, headers
    GeneLinksByIdWithHttpInfo = function(gene_ids, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_ids`)) {
        stop("Missing required parameter `gene_ids`.")
      }


      local_var_url_path <- "/gene/id/{gene_ids}/links"
      if (!missing(`gene_ids`)) {
        local_var_url_path <- gsub("\\{gene_ids\\}", paste(URLencode(as.character(`gene_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2GeneLinksReply"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene links by gene ID
    #'
    #' @param v2_gene_links_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2GeneLinksReply
    GeneLinksByIdByPost = function(v2_gene_links_request, data_file = NULL, ...) {
      local_var_response <- self$GeneLinksByIdByPostWithHttpInfo(v2_gene_links_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene links by gene ID
    #'
    #' @param v2_gene_links_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2GeneLinksReply) with additional information such as HTTP status code, headers
    GeneLinksByIdByPostWithHttpInfo = function(v2_gene_links_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_links_request`)) {
        stop("Missing required parameter `v2_gene_links_request`.")
      }


      if (!is.null(`v2_gene_links_request`)) {
        local_var_body <- `v2_gene_links_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/links"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2GeneLinksReply"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene metadata by RefSeq Accession
    #'
    #' @param accessions RNA or Protein accessions.
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneMetadataByAccession = function(accessions, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneMetadataByAccessionWithHttpInfo(accessions, returned_content, locus_tags, table_fields, include_tabular_header, page_size, page_token, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene metadata by RefSeq Accession
    #'
    #' @param accessions RNA or Protein accessions.
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneMetadataByAccessionWithHttpInfo = function(accessions, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`accessions`)) {
        stop("Missing required parameter `accessions`.")
      }












      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `locus_tags`) {
        query_params[["locus_tags"]] <- c(query_params[["locus_tags"]], list(`locus_tags` = query_item))
      }

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/accession/{accessions}"
      if (!missing(`accessions`)) {
        local_var_url_path <- gsub("\\{accessions\\}", paste(URLencode(as.character(`accessions`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene metadata as JSON
    #'
    #' @param v2_gene_dataset_reports_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneMetadataByPost = function(v2_gene_dataset_reports_request, data_file = NULL, ...) {
      local_var_response <- self$GeneMetadataByPostWithHttpInfo(v2_gene_dataset_reports_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene metadata as JSON
    #'
    #' @param v2_gene_dataset_reports_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneMetadataByPostWithHttpInfo = function(v2_gene_dataset_reports_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_dataset_reports_request`)) {
        stop("Missing required parameter `v2_gene_dataset_reports_request`.")
      }


      if (!is.null(`v2_gene_dataset_reports_request`)) {
        local_var_body <- `v2_gene_dataset_reports_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene metadata by gene symbol
    #'
    #' @param symbols Gene symbol
    #' @param taxon Taxon for provided gene symbol
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneMetadataByTaxAndSymbol = function(symbols, taxon, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneMetadataByTaxAndSymbolWithHttpInfo(symbols, taxon, returned_content, locus_tags, table_fields, include_tabular_header, page_size, page_token, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene metadata by gene symbol
    #'
    #' @param symbols Gene symbol
    #' @param taxon Taxon for provided gene symbol
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneMetadataByTaxAndSymbolWithHttpInfo = function(symbols, taxon, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`symbols`)) {
        stop("Missing required parameter `symbols`.")
      }

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }













      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `locus_tags`) {
        query_params[["locus_tags"]] <- c(query_params[["locus_tags"]], list(`locus_tags` = query_item))
      }

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/symbol/{symbols}/taxon/{taxon}"
      if (!missing(`symbols`)) {
        local_var_url_path <- gsub("\\{symbols\\}", paste(URLencode(as.character(`symbols`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene orthologs by gene ID
    #'
    #' @param gene_id 
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param taxon_filter (optional) Filter genes by taxa
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `OrthologRequest` call with more than `page_size` results. Use this token, along with the previous `OrthologRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneOrthologsById = function(gene_id, returned_content = NULL, taxon_filter = NULL, page_size = 20, page_token = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneOrthologsByIdWithHttpInfo(gene_id, returned_content, taxon_filter, page_size, page_token, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene orthologs by gene ID
    #'
    #' @param gene_id 
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param taxon_filter (optional) Filter genes by taxa
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `OrthologRequest` call with more than `page_size` results. Use this token, along with the previous `OrthologRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneOrthologsByIdWithHttpInfo = function(gene_id, returned_content = NULL, taxon_filter = NULL, page_size = 20, page_token = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_id`)) {
        stop("Missing required parameter `gene_id`.")
      }






      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `taxon_filter`) {
        query_params[["taxon_filter"]] <- c(query_params[["taxon_filter"]], list(`taxon_filter` = query_item))
      }

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      local_var_url_path <- "/gene/id/{gene_id}/orthologs"
      if (!missing(`gene_id`)) {
        local_var_url_path <- gsub("\\{gene_id\\}", URLencode(as.character(`gene_id`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene orthologs by gene ID
    #'
    #' @param v2_ortholog_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneOrthologsByPost = function(v2_ortholog_request, data_file = NULL, ...) {
      local_var_response <- self$GeneOrthologsByPostWithHttpInfo(v2_ortholog_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene orthologs by gene ID
    #'
    #' @param v2_ortholog_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneOrthologsByPostWithHttpInfo = function(v2_ortholog_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_ortholog_request`)) {
        stop("Missing required parameter `v2_ortholog_request`.")
      }


      if (!is.null(`v2_ortholog_request`)) {
        local_var_body <- `v2_ortholog_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/orthologs"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene product reports as JSON
    #'
    #' @param v2_gene_dataset_reports_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneProductReport = function(v2_gene_dataset_reports_request, data_file = NULL, ...) {
      local_var_response <- self$GeneProductReportWithHttpInfo(v2_gene_dataset_reports_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene product reports as JSON
    #'
    #' @param v2_gene_dataset_reports_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneProductReportWithHttpInfo = function(v2_gene_dataset_reports_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_gene_dataset_reports_request`)) {
        stop("Missing required parameter `v2_gene_dataset_reports_request`.")
      }


      if (!is.null(`v2_gene_dataset_reports_request`)) {
        local_var_body <- `v2_gene_dataset_reports_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/gene/product_report"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene product reports by accession IDs
    #'
    #' @param accessions RNA or Protein accessions.
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneProductReportByAccession = function(accessions, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneProductReportByAccessionWithHttpInfo(accessions, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene product reports by accession IDs
    #'
    #' @param accessions RNA or Protein accessions.
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneProductReportByAccessionWithHttpInfo = function(accessions, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`accessions`)) {
        stop("Missing required parameter `accessions`.")
      }













      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/accession/{accessions}/product_report"
      if (!missing(`accessions`)) {
        local_var_url_path <- gsub("\\{accessions\\}", paste(URLencode(as.character(`accessions`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get product reports by taxon.
    #'
    #' @param symbols Gene symbol
    #' @param taxon Taxon for provided gene symbol
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneProductReportByTaxAndSymbol = function(symbols, taxon, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneProductReportByTaxAndSymbolWithHttpInfo(symbols, taxon, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get product reports by taxon.
    #'
    #' @param symbols Gene symbol
    #' @param taxon Taxon for provided gene symbol
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneProductReportByTaxAndSymbolWithHttpInfo = function(symbols, taxon, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`symbols`)) {
        stop("Missing required parameter `symbols`.")
      }

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }














      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/symbol/{symbols}/taxon/{taxon}/product_report"
      if (!missing(`symbols`)) {
        local_var_url_path <- gsub("\\{symbols\\}", paste(URLencode(as.character(`symbols`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene product reports by gene IDs.
    #'
    #' @param gene_ids NCBI gene ids
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneProductReportsById = function(gene_ids, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneProductReportsByIdWithHttpInfo(gene_ids, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene product reports by gene IDs.
    #'
    #' @param gene_ids NCBI gene ids
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneProductReportsByIdWithHttpInfo = function(gene_ids, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_ids`)) {
        stop("Missing required parameter `gene_ids`.")
      }













      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/id/{gene_ids}/product_report"
      if (!missing(`gene_ids`)) {
        local_var_url_path <- gsub("\\{gene_ids\\}", paste(URLencode(as.character(`gene_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene product reports by locus tags
    #'
    #' @param locus_tags Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneProductReportsByLocusTags = function(locus_tags, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneProductReportsByLocusTagsWithHttpInfo(locus_tags, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene product reports by locus tags
    #'
    #' @param locus_tags Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneProductReportsByLocusTagsWithHttpInfo = function(locus_tags, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`locus_tags`)) {
        stop("Missing required parameter `locus_tags`.")
      }













      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/locus_tag/{locus_tags}/product_report"
      if (!missing(`locus_tags`)) {
        local_var_url_path <- gsub("\\{locus_tags\\}", paste(URLencode(as.character(`locus_tags`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene product reports by taxonomic identifier
    #'
    #' @param taxon NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneProductReportsByTaxon = function(taxon, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneProductReportsByTaxonWithHttpInfo(taxon, table_fields, table_format, include_tabular_header, page_size, page_token, query, types, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene product reports by taxonomic identifier
    #'
    #' @param taxon NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param table_format (optional) Optional pre-defined template for processing a tabular data request
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneProductReportsByTaxonWithHttpInfo = function(taxon, table_fields = NULL, table_format = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }













      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["table_format"]] <- `table_format`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/taxon/{taxon}/product_report"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene reports by GeneID
    #'
    #' @param gene_ids NCBI gene ids
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneReportsById = function(gene_ids, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneReportsByIdWithHttpInfo(gene_ids, returned_content, locus_tags, table_fields, include_tabular_header, page_size, page_token, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene reports by GeneID
    #'
    #' @param gene_ids NCBI gene ids
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneReportsByIdWithHttpInfo = function(gene_ids, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`gene_ids`)) {
        stop("Missing required parameter `gene_ids`.")
      }












      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `locus_tags`) {
        query_params[["locus_tags"]] <- c(query_params[["locus_tags"]], list(`locus_tags` = query_item))
      }

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/id/{gene_ids}"
      if (!missing(`gene_ids`)) {
        local_var_url_path <- gsub("\\{gene_ids\\}", paste(URLencode(as.character(`gene_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get gene reports by taxonomic identifier
    #'
    #' @param taxon NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsGeneDataReportPage
    GeneReportsByTaxon = function(taxon, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      local_var_response <- self$GeneReportsByTaxonWithHttpInfo(taxon, returned_content, locus_tags, table_fields, include_tabular_header, page_size, page_token, query, types, accession_filter, tax_search_subtree, sort_field, sort_direction, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get gene reports by taxonomic identifier
    #'
    #' @param taxon NCBI Taxonomy ID or name (common or scientific) that the genes are annotated at
    #' @param returned_content (optional) Return either gene-ids, or entire gene metadata
    #' @param locus_tags (optional) Gene locus tags
    #' @param table_fields (optional) Specify which fields to include in the tabular report
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_size (optional) The maximum number of gene reports to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from an `AssemblyDatasetReportsRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyDatasetReportsRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param query (optional) text search within gene symbol, aliases, name, locus-tag and protein name
    #' @param types (optional) Gene types to filter
    #' @param accession_filter (optional) Filter tabular product report to only include the selected accessions
    #' @param tax_search_subtree (optional) For queries including a tax-id, include any matching genes annotated on taxa below the selected taxon. By default this is false, and only genes for the selected taxon are returned. (default value: FALSE)
    #' @param sort_field (optional) No description
    #' @param sort_direction (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsGeneDataReportPage) with additional information such as HTTP status code, headers
    GeneReportsByTaxonWithHttpInfo = function(taxon, returned_content = NULL, locus_tags = NULL, table_fields = NULL, include_tabular_header = NULL, page_size = 20, page_token = NULL, query = NULL, types = NULL, accession_filter = NULL, tax_search_subtree = FALSE, sort_field = NULL, sort_direction = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }














      query_params[["returned_content"]] <- `returned_content`

      # explore
      for (query_item in `locus_tags`) {
        query_params[["locus_tags"]] <- c(query_params[["locus_tags"]], list(`locus_tags` = query_item))
      }

      # explore
      for (query_item in `table_fields`) {
        query_params[["table_fields"]] <- c(query_params[["table_fields"]], list(`table_fields` = query_item))
      }

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      query_params[["query"]] <- `query`

      # explore
      for (query_item in `types`) {
        query_params[["types"]] <- c(query_params[["types"]], list(`types` = query_item))
      }

      # explore
      for (query_item in `accession_filter`) {
        query_params[["accession_filter"]] <- c(query_params[["accession_filter"]], list(`accession_filter` = query_item))
      }

      query_params[["tax_search_subtree"]] <- `tax_search_subtree`

      query_params[["sort.field"]] <- `sort_field`

      query_params[["sort.direction"]] <- `sort_direction`

      local_var_url_path <- "/gene/taxon/{taxon}"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsGeneDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    }
  )
)
