#' NCBI Datasets API
#'
#' ### NCBI Datasets is a resource that lets you easily gather data from NCBI. The Datasets version 2 API is still in alpha, and we're updating it often to add new functionality, iron out bugs and enhance usability. For some larger downloads, you may want to download a [dehydrated zip archive](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/how-tos/genomes/large-download/), and retrieve the individual data files at a later time. 
#'
#' The version of the OpenAPI document: v2
#' Generated by: https://openapi-generator.tech
#'
#' @docType class
#' @title Taxonomy operations
#' @description TaxonomyApi
#' @format An \code{R6Class} generator object
#' @field api_client Handles the client-server communication.
#'
#' @examples
#' \dontrun{
#' ####################  DownloadTaxonomyPackage  ####################
#'
#' library(ncbiopenapi)
#' var_tax_ids <- c(123) # array[integer] | 
#' var_aux_reports <- c(v2TaxonomyDatasetRequestTaxonomyReportType$new()) # array[V2TaxonomyDatasetRequestTaxonomyReportType] | list additional reports to include with download. TAXONOMY_REPORT is included by default. (Optional)
#' var_filename <- "ncbi_dataset.zip" # character | Output file name. (Optional)
#'
#' #Get a taxonomy data package by tax ID
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$DownloadTaxonomyPackage(var_tax_ids, aux_reports = var_aux_reports, filename = var_filenamedata_file = "result.txt")
#' result <- api_instance$DownloadTaxonomyPackage(var_tax_ids, aux_reports = var_aux_reports, filename = var_filename)
#' dput(result)
#'
#'
#' ####################  DownloadTaxonomyPackageByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_dataset_request <- v2TaxonomyDatasetRequest$new(c(123), c(v2TaxonomyDatasetRequestTaxonomyReportType$new())) # V2TaxonomyDatasetRequest | 
#' var_filename <- "ncbi_dataset.zip" # character | Output file name. (Optional)
#'
#' #Get a taxonomy data package by tax_id
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$DownloadTaxonomyPackageByPost(var_v2_taxonomy_dataset_request, filename = var_filenamedata_file = "result.txt")
#' result <- api_instance$DownloadTaxonomyPackageByPost(var_v2_taxonomy_dataset_request, filename = var_filename)
#' dput(result)
#'
#'
#' ####################  TaxNameQuery  ####################
#'
#' library(ncbiopenapi)
#' var_taxon_query <- "hum" # character | NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
#' var_tax_rank_filter <- v2OrganismQueryRequestTaxRankFilter$new() # V2OrganismQueryRequestTaxRankFilter | Set the scope of searched tax ranks when filtering by gene or genome.  Not used for 'all' (Optional)
#' var_taxon_resource_filter <- v2OrganismQueryRequestTaxonResourceFilter$new() # V2OrganismQueryRequestTaxonResourceFilter | Limit results to those with gene or genome counts (no filter by default) (Optional)
#' var_exact_match <- FALSE # character | If true, only return results that exactly match the provided name or tax-id (Optional)
#'
#' #Get a list of taxonomy names and IDs given a partial taxonomic name
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxNameQuery(var_taxon_query, tax_rank_filter = var_tax_rank_filter, taxon_resource_filter = var_taxon_resource_filter, exact_match = var_exact_matchdata_file = "result.txt")
#' result <- api_instance$TaxNameQuery(var_taxon_query, tax_rank_filter = var_tax_rank_filter, taxon_resource_filter = var_taxon_resource_filter, exact_match = var_exact_match)
#' dput(result)
#'
#'
#' ####################  TaxNameQueryByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_organism_query_request <- v2OrganismQueryRequest$new("organism_query_example", "taxon_query_example", v2OrganismQueryRequestTaxRankFilter$new(), v2OrganismQueryRequestTaxonResourceFilter$new(), "exact_match_example") # V2OrganismQueryRequest | 
#'
#' #Get a list of taxonomy names and IDs given a partial taxonomic name
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxNameQueryByPost(var_v2_organism_query_requestdata_file = "result.txt")
#' result <- api_instance$TaxNameQueryByPost(var_v2_organism_query_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyDataReport  ####################
#'
#' library(ncbiopenapi)
#' var_taxons <- c("inner_example") # array[character] | 
#' var_returned_content <- v2TaxonomyMetadataRequestContentType$new() # V2TaxonomyMetadataRequestContentType | Return either tax-ids alone, or entire taxononmy-metadata records (Optional)
#' var_page_size <- 20 # integer | The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_table_format <- v2TaxonomyMetadataRequestTableFormat$new() # V2TaxonomyMetadataRequestTableFormat |  (Optional)
#' var_children <- "children_example" # character | Flag for tax explosion. (Optional)
#' var_ranks <- c(v2reportsRankType$new()) # array[V2reportsRankType] | Only include taxons of the provided ranks. If empty, return all ranks. (Optional)
#'
#' #Use taxonomic identifiers to get taxonomic data report
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyDataReport(var_taxons, returned_content = var_returned_content, page_size = var_page_size, include_tabular_header = var_include_tabular_header, page_token = var_page_token, table_format = var_table_format, children = var_children, ranks = var_ranksdata_file = "result.txt")
#' result <- api_instance$TaxonomyDataReport(var_taxons, returned_content = var_returned_content, page_size = var_page_size, include_tabular_header = var_include_tabular_header, page_token = var_page_token, table_format = var_table_format, children = var_children, ranks = var_ranks)
#' dput(result)
#'
#'
#' ####################  TaxonomyDataReportPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_metadata_request <- v2TaxonomyMetadataRequest$new(c("taxons_example"), v2TaxonomyMetadataRequestContentType$new(), 123, v2IncludeTabularHeader$new(), "page_token_example", v2TaxonomyMetadataRequestTableFormat$new(), "children_example", c(v2reportsRankType$new())) # V2TaxonomyMetadataRequest | 
#'
#' #Use taxonomic identifiers to get taxonomic names data report by post
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyDataReportPost(var_v2_taxonomy_metadata_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyDataReportPost(var_v2_taxonomy_metadata_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyFilteredSubtree  ####################
#'
#' library(ncbiopenapi)
#' var_taxons <- c("inner_example") # array[character] | 
#' var_specified_limit <- "specified_limit_example" # character | Limit to specified species (Optional)
#' var_rank_limits <- c(v2reportsRankType$new()) # array[V2reportsRankType] | Limit to the provided ranks.  If empty, accept any rank. (Optional)
#'
#' #Use taxonomic identifiers to get a filtered taxonomic subtree
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyFilteredSubtree(var_taxons, specified_limit = var_specified_limit, rank_limits = var_rank_limitsdata_file = "result.txt")
#' result <- api_instance$TaxonomyFilteredSubtree(var_taxons, specified_limit = var_specified_limit, rank_limits = var_rank_limits)
#' dput(result)
#'
#'
#' ####################  TaxonomyFilteredSubtreePost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_filtered_subtree_request <- v2TaxonomyFilteredSubtreeRequest$new(c("taxons_example"), "specified_limit_example", c(v2reportsRankType$new())) # V2TaxonomyFilteredSubtreeRequest | 
#'
#' #Use taxonomic identifiers to get a filtered taxonomic subtree by post
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyFilteredSubtreePost(var_v2_taxonomy_filtered_subtree_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyFilteredSubtreePost(var_v2_taxonomy_filtered_subtree_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyImage  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | 
#' var_image_size <- v2ImageSize$new() # V2ImageSize |  (Optional)
#'
#' #Retrieve image associated with a taxonomic identifier
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyImage(var_taxon, image_size = var_image_sizedata_file = "result.txt")
#' result <- api_instance$TaxonomyImage(var_taxon, image_size = var_image_size)
#' dput(result)
#'
#'
#' ####################  TaxonomyImageMetadata  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | 
#'
#' #Retrieve image metadata associated with a taxonomic identifier
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyImageMetadata(var_taxondata_file = "result.txt")
#' result <- api_instance$TaxonomyImageMetadata(var_taxon)
#' dput(result)
#'
#'
#' ####################  TaxonomyImageMetadataPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_image_metadata_request <- v2TaxonomyImageMetadataRequest$new("taxon_example") # V2TaxonomyImageMetadataRequest | 
#'
#' #Retrieve image metadata associated with a taxonomic identifier by post
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyImageMetadataPost(var_v2_taxonomy_image_metadata_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyImageMetadataPost(var_v2_taxonomy_image_metadata_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyImagePost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_image_request <- v2TaxonomyImageRequest$new("taxon_example", v2ImageSize$new()) # V2TaxonomyImageRequest | 
#'
#' #Retrieve image associated with a taxonomic identifier by post
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyImagePost(var_v2_taxonomy_image_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyImagePost(var_v2_taxonomy_image_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyLinks  ####################
#'
#' library(ncbiopenapi)
#' var_taxon <- "9606" # character | 
#'
#' #Retrieve external links associated with a taxonomic identifier.
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyLinks(var_taxondata_file = "result.txt")
#' result <- api_instance$TaxonomyLinks(var_taxon)
#' dput(result)
#'
#'
#' ####################  TaxonomyLinksByPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_links_request <- v2TaxonomyLinksRequest$new("taxon_example") # V2TaxonomyLinksRequest | 
#'
#' #Retrieve external links associated with a taxonomic identifier.
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyLinksByPost(var_v2_taxonomy_links_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyLinksByPost(var_v2_taxonomy_links_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyMetadata  ####################
#'
#' library(ncbiopenapi)
#' var_taxons <- c("inner_example") # array[character] | 
#' var_returned_content <- v2TaxonomyMetadataRequestContentType$new() # V2TaxonomyMetadataRequestContentType | Return either tax-ids alone, or entire taxononmy-metadata records (Optional)
#' var_page_size <- 20 # integer | The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_table_format <- v2TaxonomyMetadataRequestTableFormat$new() # V2TaxonomyMetadataRequestTableFormat |  (Optional)
#' var_children <- "children_example" # character | Flag for tax explosion. (Optional)
#' var_ranks <- c(v2reportsRankType$new()) # array[V2reportsRankType] | Only include taxons of the provided ranks. If empty, return all ranks. (Optional)
#'
#' #Use taxonomic identifiers to get taxonomic metadata
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyMetadata(var_taxons, returned_content = var_returned_content, page_size = var_page_size, include_tabular_header = var_include_tabular_header, page_token = var_page_token, table_format = var_table_format, children = var_children, ranks = var_ranksdata_file = "result.txt")
#' result <- api_instance$TaxonomyMetadata(var_taxons, returned_content = var_returned_content, page_size = var_page_size, include_tabular_header = var_include_tabular_header, page_token = var_page_token, table_format = var_table_format, children = var_children, ranks = var_ranks)
#' dput(result)
#'
#'
#' ####################  TaxonomyMetadataPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_metadata_request <- v2TaxonomyMetadataRequest$new(c("taxons_example"), v2TaxonomyMetadataRequestContentType$new(), 123, v2IncludeTabularHeader$new(), "page_token_example", v2TaxonomyMetadataRequestTableFormat$new(), "children_example", c(v2reportsRankType$new())) # V2TaxonomyMetadataRequest | 
#'
#' #Use taxonomic identifiers to get taxonomic metadata by post
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyMetadataPost(var_v2_taxonomy_metadata_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyMetadataPost(var_v2_taxonomy_metadata_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyNames  ####################
#'
#' library(ncbiopenapi)
#' var_taxons <- c("inner_example") # array[character] | 
#' var_returned_content <- v2TaxonomyMetadataRequestContentType$new() # V2TaxonomyMetadataRequestContentType | Return either tax-ids alone, or entire taxononmy-metadata records (Optional)
#' var_page_size <- 20 # integer | The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_include_tabular_header <- v2IncludeTabularHeader$new() # V2IncludeTabularHeader | Whether this request for tabular data should include the header row (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved. (Optional)
#' var_table_format <- v2TaxonomyMetadataRequestTableFormat$new() # V2TaxonomyMetadataRequestTableFormat |  (Optional)
#' var_children <- "children_example" # character | Flag for tax explosion. (Optional)
#' var_ranks <- c(v2reportsRankType$new()) # array[V2reportsRankType] | Only include taxons of the provided ranks. If empty, return all ranks. (Optional)
#'
#' #Use taxonomic identifiers to get taxonomic names data report
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyNames(var_taxons, returned_content = var_returned_content, page_size = var_page_size, include_tabular_header = var_include_tabular_header, page_token = var_page_token, table_format = var_table_format, children = var_children, ranks = var_ranksdata_file = "result.txt")
#' result <- api_instance$TaxonomyNames(var_taxons, returned_content = var_returned_content, page_size = var_page_size, include_tabular_header = var_include_tabular_header, page_token = var_page_token, table_format = var_table_format, children = var_children, ranks = var_ranks)
#' dput(result)
#'
#'
#' ####################  TaxonomyNamesPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_metadata_request <- v2TaxonomyMetadataRequest$new(c("taxons_example"), v2TaxonomyMetadataRequestContentType$new(), 123, v2IncludeTabularHeader$new(), "page_token_example", v2TaxonomyMetadataRequestTableFormat$new(), "children_example", c(v2reportsRankType$new())) # V2TaxonomyMetadataRequest | 
#'
#' #Use taxonomic identifiers to get taxonomic names data report by post
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyNamesPost(var_v2_taxonomy_metadata_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyNamesPost(var_v2_taxonomy_metadata_request)
#' dput(result)
#'
#'
#' ####################  TaxonomyRelatedIds  ####################
#'
#' library(ncbiopenapi)
#' var_tax_id <- 9606 # integer | 
#' var_include_lineage <- FALSE # character | If true, return reports for all taxonomy nodes in the lineages of the requested tax_id (Optional)
#' var_include_subtree <- FALSE # character | This field is deprecated because all requests include the subtree, so it has no effect (Optional)
#' var_ranks <- c(v2reportsRankType$new()) # array[V2reportsRankType] | Only include taxons of the provided ranks. If empty, return all ranks. (Optional)
#' var_page_size <- 20 # integer | The maximum number of taxids to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (Optional)
#' var_page_token <- "page_token_example" # character | A page token is returned from a `GetRelatedTaxids` call with more than `page_size` results. Use this token, along with the previous `TaxonomyRelatedIdRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved. (Optional)
#'
#' #Use taxonomic identifier to get related taxonomic identifiers, such as children
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyRelatedIds(var_tax_id, include_lineage = var_include_lineage, include_subtree = var_include_subtree, ranks = var_ranks, page_size = var_page_size, page_token = var_page_tokendata_file = "result.txt")
#' result <- api_instance$TaxonomyRelatedIds(var_tax_id, include_lineage = var_include_lineage, include_subtree = var_include_subtree, ranks = var_ranks, page_size = var_page_size, page_token = var_page_token)
#' dput(result)
#'
#'
#' ####################  TaxonomyRelatedIdsPost  ####################
#'
#' library(ncbiopenapi)
#' var_v2_taxonomy_related_id_request <- v2TaxonomyRelatedIdRequest$new(123, "include_lineage_example", "include_subtree_example", c(v2reportsRankType$new()), 123, "page_token_example") # V2TaxonomyRelatedIdRequest | 
#'
#' #Use taxonomic identifier to get related taxonomic identifiers, such as children
#' api_instance <- TaxonomyApi$new()
#'
#' # Configure API key authorization: ApiKeyAuthHeader
#' api_instance$api_client$api_keys["api-key"] <- Sys.getenv("API_KEY")
#'
#' # to save the result into a file, simply add the optional `data_file` parameter, e.g.
#' # result <- api_instance$TaxonomyRelatedIdsPost(var_v2_taxonomy_related_id_requestdata_file = "result.txt")
#' result <- api_instance$TaxonomyRelatedIdsPost(var_v2_taxonomy_related_id_request)
#' dput(result)
#'
#'
#' }
#' @importFrom R6 R6Class
#' @importFrom base64enc base64encode
#' @export
TaxonomyApi <- R6::R6Class(
  "TaxonomyApi",
  public = list(
    api_client = NULL,

    #' @description
    #' Initialize a new TaxonomyApi.
    #'
    #' @param api_client An instance of API client.
    initialize = function(api_client) {
      if (!missing(api_client)) {
        self$api_client <- api_client
      } else {
        self$api_client <- ApiClient$new()
      }
    },

    #' @description
    #' Get a taxonomy data package by tax ID
    #'
    #' @param tax_ids 
    #' @param aux_reports (optional) list additional reports to include with download. TAXONOMY_REPORT is included by default.
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return data.frame
    DownloadTaxonomyPackage = function(tax_ids, aux_reports = NULL, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      local_var_response <- self$DownloadTaxonomyPackageWithHttpInfo(tax_ids, aux_reports, filename, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get a taxonomy data package by tax ID
    #'
    #' @param tax_ids 
    #' @param aux_reports (optional) list additional reports to include with download. TAXONOMY_REPORT is included by default.
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (data.frame) with additional information such as HTTP status code, headers
    DownloadTaxonomyPackageWithHttpInfo = function(tax_ids, aux_reports = NULL, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`tax_ids`)) {
        stop("Missing required parameter `tax_ids`.")
      }




      # explore
      for (query_item in `aux_reports`) {
        query_params[["aux_reports"]] <- c(query_params[["aux_reports"]], list(`aux_reports` = query_item))
      }

      query_params[["filename"]] <- `filename`

      local_var_url_path <- "/taxonomy/taxon/{tax_ids}/download"
      if (!missing(`tax_ids`)) {
        local_var_url_path <- gsub("\\{tax_ids\\}", paste(URLencode(as.character(`tax_ids`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/zip")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "data.frame"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get a taxonomy data package by tax_id
    #'
    #' @param v2_taxonomy_dataset_request 
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return data.frame
    DownloadTaxonomyPackageByPost = function(v2_taxonomy_dataset_request, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      local_var_response <- self$DownloadTaxonomyPackageByPostWithHttpInfo(v2_taxonomy_dataset_request, filename, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get a taxonomy data package by tax_id
    #'
    #' @param v2_taxonomy_dataset_request 
    #' @param filename (optional) Output file name. (default value: "ncbi_dataset.zip")
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (data.frame) with additional information such as HTTP status code, headers
    DownloadTaxonomyPackageByPostWithHttpInfo = function(v2_taxonomy_dataset_request, filename = "ncbi_dataset.zip", data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_dataset_request`)) {
        stop("Missing required parameter `v2_taxonomy_dataset_request`.")
      }



      query_params[["filename"]] <- `filename`

      if (!is.null(`v2_taxonomy_dataset_request`)) {
        local_var_body <- `v2_taxonomy_dataset_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/download"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/zip")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "data.frame"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get a list of taxonomy names and IDs given a partial taxonomic name
    #'
    #' @param taxon_query NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
    #' @param tax_rank_filter (optional) Set the scope of searched tax ranks when filtering by gene or genome.  Not used for 'all'
    #' @param taxon_resource_filter (optional) Limit results to those with gene or genome counts (no filter by default)
    #' @param exact_match (optional) If true, only return results that exactly match the provided name or tax-id (default value: FALSE)
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2SciNameAndIds
    TaxNameQuery = function(taxon_query, tax_rank_filter = NULL, taxon_resource_filter = NULL, exact_match = FALSE, data_file = NULL, ...) {
      local_var_response <- self$TaxNameQueryWithHttpInfo(taxon_query, tax_rank_filter, taxon_resource_filter, exact_match, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get a list of taxonomy names and IDs given a partial taxonomic name
    #'
    #' @param taxon_query NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
    #' @param tax_rank_filter (optional) Set the scope of searched tax ranks when filtering by gene or genome.  Not used for 'all'
    #' @param taxon_resource_filter (optional) Limit results to those with gene or genome counts (no filter by default)
    #' @param exact_match (optional) If true, only return results that exactly match the provided name or tax-id (default value: FALSE)
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2SciNameAndIds) with additional information such as HTTP status code, headers
    TaxNameQueryWithHttpInfo = function(taxon_query, tax_rank_filter = NULL, taxon_resource_filter = NULL, exact_match = FALSE, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon_query`)) {
        stop("Missing required parameter `taxon_query`.")
      }





      query_params[["tax_rank_filter"]] <- `tax_rank_filter`

      query_params[["taxon_resource_filter"]] <- `taxon_resource_filter`

      query_params[["exact_match"]] <- `exact_match`

      local_var_url_path <- "/taxonomy/taxon_suggest/{taxon_query}"
      if (!missing(`taxon_query`)) {
        local_var_url_path <- gsub("\\{taxon_query\\}", URLencode(as.character(`taxon_query`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2SciNameAndIds"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Get a list of taxonomy names and IDs given a partial taxonomic name
    #'
    #' @param v2_organism_query_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2SciNameAndIds
    TaxNameQueryByPost = function(v2_organism_query_request, data_file = NULL, ...) {
      local_var_response <- self$TaxNameQueryByPostWithHttpInfo(v2_organism_query_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Get a list of taxonomy names and IDs given a partial taxonomic name
    #'
    #' @param v2_organism_query_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2SciNameAndIds) with additional information such as HTTP status code, headers
    TaxNameQueryByPostWithHttpInfo = function(v2_organism_query_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_organism_query_request`)) {
        stop("Missing required parameter `v2_organism_query_request`.")
      }


      if (!is.null(`v2_organism_query_request`)) {
        local_var_body <- `v2_organism_query_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/taxon_suggest"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2SciNameAndIds"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic data report
    #'
    #' @param taxons 
    #' @param returned_content (optional) Return either tax-ids alone, or entire taxononmy-metadata records
    #' @param page_size (optional) The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_token (optional) A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved.
    #' @param table_format (optional) No description
    #' @param children (optional) Flag for tax explosion.
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsTaxonomyDataReportPage
    TaxonomyDataReport = function(taxons, returned_content = NULL, page_size = 20, include_tabular_header = NULL, page_token = NULL, table_format = NULL, children = NULL, ranks = NULL, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyDataReportWithHttpInfo(taxons, returned_content, page_size, include_tabular_header, page_token, table_format, children, ranks, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic data report
    #'
    #' @param taxons 
    #' @param returned_content (optional) Return either tax-ids alone, or entire taxononmy-metadata records
    #' @param page_size (optional) The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_token (optional) A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved.
    #' @param table_format (optional) No description
    #' @param children (optional) Flag for tax explosion.
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsTaxonomyDataReportPage) with additional information such as HTTP status code, headers
    TaxonomyDataReportWithHttpInfo = function(taxons, returned_content = NULL, page_size = 20, include_tabular_header = NULL, page_token = NULL, table_format = NULL, children = NULL, ranks = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxons`)) {
        stop("Missing required parameter `taxons`.")
      }









      query_params[["returned_content"]] <- `returned_content`

      query_params[["page_size"]] <- `page_size`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_token"]] <- `page_token`

      query_params[["table_format"]] <- `table_format`

      query_params[["children"]] <- `children`

      # explore
      for (query_item in `ranks`) {
        query_params[["ranks"]] <- c(query_params[["ranks"]], list(`ranks` = query_item))
      }

      local_var_url_path <- "/taxonomy/taxon/{taxons}/dataset_report"
      if (!missing(`taxons`)) {
        local_var_url_path <- gsub("\\{taxons\\}", paste(URLencode(as.character(`taxons`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsTaxonomyDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic names data report by post
    #'
    #' @param v2_taxonomy_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsTaxonomyDataReportPage
    TaxonomyDataReportPost = function(v2_taxonomy_metadata_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyDataReportPostWithHttpInfo(v2_taxonomy_metadata_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic names data report by post
    #'
    #' @param v2_taxonomy_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsTaxonomyDataReportPage) with additional information such as HTTP status code, headers
    TaxonomyDataReportPostWithHttpInfo = function(v2_taxonomy_metadata_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_metadata_request`)) {
        stop("Missing required parameter `v2_taxonomy_metadata_request`.")
      }


      if (!is.null(`v2_taxonomy_metadata_request`)) {
        local_var_body <- `v2_taxonomy_metadata_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/dataset_report"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json", "application/x-ndjson", "text/tab-separated-values")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsTaxonomyDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get a filtered taxonomic subtree
    #'
    #' @param taxons 
    #' @param specified_limit (optional) Limit to specified species
    #' @param rank_limits (optional) Limit to the provided ranks.  If empty, accept any rank.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyFilteredSubtreeResponse
    TaxonomyFilteredSubtree = function(taxons, specified_limit = NULL, rank_limits = NULL, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyFilteredSubtreeWithHttpInfo(taxons, specified_limit, rank_limits, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get a filtered taxonomic subtree
    #'
    #' @param taxons 
    #' @param specified_limit (optional) Limit to specified species
    #' @param rank_limits (optional) Limit to the provided ranks.  If empty, accept any rank.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyFilteredSubtreeResponse) with additional information such as HTTP status code, headers
    TaxonomyFilteredSubtreeWithHttpInfo = function(taxons, specified_limit = NULL, rank_limits = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxons`)) {
        stop("Missing required parameter `taxons`.")
      }




      query_params[["specified_limit"]] <- `specified_limit`

      # explore
      for (query_item in `rank_limits`) {
        query_params[["rank_limits"]] <- c(query_params[["rank_limits"]], list(`rank_limits` = query_item))
      }

      local_var_url_path <- "/taxonomy/taxon/{taxons}/filtered_subtree"
      if (!missing(`taxons`)) {
        local_var_url_path <- gsub("\\{taxons\\}", paste(URLencode(as.character(`taxons`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyFilteredSubtreeResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get a filtered taxonomic subtree by post
    #'
    #' @param v2_taxonomy_filtered_subtree_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyFilteredSubtreeResponse
    TaxonomyFilteredSubtreePost = function(v2_taxonomy_filtered_subtree_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyFilteredSubtreePostWithHttpInfo(v2_taxonomy_filtered_subtree_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get a filtered taxonomic subtree by post
    #'
    #' @param v2_taxonomy_filtered_subtree_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyFilteredSubtreeResponse) with additional information such as HTTP status code, headers
    TaxonomyFilteredSubtreePostWithHttpInfo = function(v2_taxonomy_filtered_subtree_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_filtered_subtree_request`)) {
        stop("Missing required parameter `v2_taxonomy_filtered_subtree_request`.")
      }


      if (!is.null(`v2_taxonomy_filtered_subtree_request`)) {
        local_var_body <- `v2_taxonomy_filtered_subtree_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/filtered_subtree"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyFilteredSubtreeResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Retrieve image associated with a taxonomic identifier
    #'
    #' @param taxon 
    #' @param image_size (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return data.frame
    TaxonomyImage = function(taxon, image_size = NULL, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyImageWithHttpInfo(taxon, image_size, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Retrieve image associated with a taxonomic identifier
    #'
    #' @param taxon 
    #' @param image_size (optional) No description
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (data.frame) with additional information such as HTTP status code, headers
    TaxonomyImageWithHttpInfo = function(taxon, image_size = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }



      query_params[["image_size"]] <- `image_size`

      local_var_url_path <- "/taxonomy/taxon/{taxon}/image"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "image/jpeg", "image/png", "image/gif", "image/tiff", "image/svg+xml")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "data.frame"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Retrieve image metadata associated with a taxonomic identifier
    #'
    #' @param taxon 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyImageMetadataResponse
    TaxonomyImageMetadata = function(taxon, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyImageMetadataWithHttpInfo(taxon, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Retrieve image metadata associated with a taxonomic identifier
    #'
    #' @param taxon 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyImageMetadataResponse) with additional information such as HTTP status code, headers
    TaxonomyImageMetadataWithHttpInfo = function(taxon, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }


      local_var_url_path <- "/taxonomy/taxon/{taxon}/image/metadata"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyImageMetadataResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Retrieve image metadata associated with a taxonomic identifier by post
    #'
    #' @param v2_taxonomy_image_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyImageMetadataResponse
    TaxonomyImageMetadataPost = function(v2_taxonomy_image_metadata_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyImageMetadataPostWithHttpInfo(v2_taxonomy_image_metadata_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Retrieve image metadata associated with a taxonomic identifier by post
    #'
    #' @param v2_taxonomy_image_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyImageMetadataResponse) with additional information such as HTTP status code, headers
    TaxonomyImageMetadataPostWithHttpInfo = function(v2_taxonomy_image_metadata_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_image_metadata_request`)) {
        stop("Missing required parameter `v2_taxonomy_image_metadata_request`.")
      }


      if (!is.null(`v2_taxonomy_image_metadata_request`)) {
        local_var_body <- `v2_taxonomy_image_metadata_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/image/metadata"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyImageMetadataResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Retrieve image associated with a taxonomic identifier by post
    #'
    #' @param v2_taxonomy_image_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return data.frame
    TaxonomyImagePost = function(v2_taxonomy_image_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyImagePostWithHttpInfo(v2_taxonomy_image_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Retrieve image associated with a taxonomic identifier by post
    #'
    #' @param v2_taxonomy_image_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (data.frame) with additional information such as HTTP status code, headers
    TaxonomyImagePostWithHttpInfo = function(v2_taxonomy_image_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_image_request`)) {
        stop("Missing required parameter `v2_taxonomy_image_request`.")
      }


      if (!is.null(`v2_taxonomy_image_request`)) {
        local_var_body <- `v2_taxonomy_image_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/image"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "image/jpeg", "image/png", "image/tiff", "image/svg+xml")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "data.frame"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Retrieve external links associated with a taxonomic identifier.
    #'
    #' @param taxon 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyLinksResponse
    TaxonomyLinks = function(taxon, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyLinksWithHttpInfo(taxon, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Retrieve external links associated with a taxonomic identifier.
    #'
    #' @param taxon 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyLinksResponse) with additional information such as HTTP status code, headers
    TaxonomyLinksWithHttpInfo = function(taxon, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }


      local_var_url_path <- "/taxonomy/taxon/{taxon}/links"
      if (!missing(`taxon`)) {
        local_var_url_path <- gsub("\\{taxon\\}", URLencode(as.character(`taxon`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyLinksResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Retrieve external links associated with a taxonomic identifier.
    #'
    #' @param v2_taxonomy_links_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyLinksResponse
    TaxonomyLinksByPost = function(v2_taxonomy_links_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyLinksByPostWithHttpInfo(v2_taxonomy_links_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Retrieve external links associated with a taxonomic identifier.
    #'
    #' @param v2_taxonomy_links_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyLinksResponse) with additional information such as HTTP status code, headers
    TaxonomyLinksByPostWithHttpInfo = function(v2_taxonomy_links_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_links_request`)) {
        stop("Missing required parameter `v2_taxonomy_links_request`.")
      }


      if (!is.null(`v2_taxonomy_links_request`)) {
        local_var_body <- `v2_taxonomy_links_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/links"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyLinksResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic metadata
    #'
    #' @param taxons 
    #' @param returned_content (optional) Return either tax-ids alone, or entire taxononmy-metadata records
    #' @param page_size (optional) The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_token (optional) A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved.
    #' @param table_format (optional) No description
    #' @param children (optional) Flag for tax explosion.
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyMetadataResponse
    TaxonomyMetadata = function(taxons, returned_content = NULL, page_size = 20, include_tabular_header = NULL, page_token = NULL, table_format = NULL, children = NULL, ranks = NULL, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyMetadataWithHttpInfo(taxons, returned_content, page_size, include_tabular_header, page_token, table_format, children, ranks, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic metadata
    #'
    #' @param taxons 
    #' @param returned_content (optional) Return either tax-ids alone, or entire taxononmy-metadata records
    #' @param page_size (optional) The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_token (optional) A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved.
    #' @param table_format (optional) No description
    #' @param children (optional) Flag for tax explosion.
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyMetadataResponse) with additional information such as HTTP status code, headers
    TaxonomyMetadataWithHttpInfo = function(taxons, returned_content = NULL, page_size = 20, include_tabular_header = NULL, page_token = NULL, table_format = NULL, children = NULL, ranks = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxons`)) {
        stop("Missing required parameter `taxons`.")
      }









      query_params[["returned_content"]] <- `returned_content`

      query_params[["page_size"]] <- `page_size`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_token"]] <- `page_token`

      query_params[["table_format"]] <- `table_format`

      query_params[["children"]] <- `children`

      # explore
      for (query_item in `ranks`) {
        query_params[["ranks"]] <- c(query_params[["ranks"]], list(`ranks` = query_item))
      }

      local_var_url_path <- "/taxonomy/taxon/{taxons}"
      if (!missing(`taxons`)) {
        local_var_url_path <- gsub("\\{taxons\\}", paste(URLencode(as.character(`taxons`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyMetadataResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic metadata by post
    #'
    #' @param v2_taxonomy_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyMetadataResponse
    TaxonomyMetadataPost = function(v2_taxonomy_metadata_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyMetadataPostWithHttpInfo(v2_taxonomy_metadata_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic metadata by post
    #'
    #' @param v2_taxonomy_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyMetadataResponse) with additional information such as HTTP status code, headers
    TaxonomyMetadataPostWithHttpInfo = function(v2_taxonomy_metadata_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_metadata_request`)) {
        stop("Missing required parameter `v2_taxonomy_metadata_request`.")
      }


      if (!is.null(`v2_taxonomy_metadata_request`)) {
        local_var_body <- `v2_taxonomy_metadata_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyMetadataResponse"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic names data report
    #'
    #' @param taxons 
    #' @param returned_content (optional) Return either tax-ids alone, or entire taxononmy-metadata records
    #' @param page_size (optional) The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_token (optional) A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved.
    #' @param table_format (optional) No description
    #' @param children (optional) Flag for tax explosion.
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsTaxonomyNamesDataReportPage
    TaxonomyNames = function(taxons, returned_content = NULL, page_size = 20, include_tabular_header = NULL, page_token = NULL, table_format = NULL, children = NULL, ranks = NULL, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyNamesWithHttpInfo(taxons, returned_content, page_size, include_tabular_header, page_token, table_format, children, ranks, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic names data report
    #'
    #' @param taxons 
    #' @param returned_content (optional) Return either tax-ids alone, or entire taxononmy-metadata records
    #' @param page_size (optional) The maximum number of taxons to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param include_tabular_header (optional) Whether this request for tabular data should include the header row
    #' @param page_token (optional) A page token is returned from `GetTaxonomyDataReportFor` and `GetTaxonomyNamesDataReportFor` calls with more than `page_size` results. When `page_token` is empty, all results have been retrieved.
    #' @param table_format (optional) No description
    #' @param children (optional) Flag for tax explosion.
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsTaxonomyNamesDataReportPage) with additional information such as HTTP status code, headers
    TaxonomyNamesWithHttpInfo = function(taxons, returned_content = NULL, page_size = 20, include_tabular_header = NULL, page_token = NULL, table_format = NULL, children = NULL, ranks = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`taxons`)) {
        stop("Missing required parameter `taxons`.")
      }









      query_params[["returned_content"]] <- `returned_content`

      query_params[["page_size"]] <- `page_size`

      query_params[["include_tabular_header"]] <- `include_tabular_header`

      query_params[["page_token"]] <- `page_token`

      query_params[["table_format"]] <- `table_format`

      query_params[["children"]] <- `children`

      # explore
      for (query_item in `ranks`) {
        query_params[["ranks"]] <- c(query_params[["ranks"]], list(`ranks` = query_item))
      }

      local_var_url_path <- "/taxonomy/taxon/{taxons}/name_report"
      if (!missing(`taxons`)) {
        local_var_url_path <- gsub("\\{taxons\\}", paste(URLencode(as.character(`taxons`), reserved = TRUE), collapse= ",", sep=""), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsTaxonomyNamesDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic names data report by post
    #'
    #' @param v2_taxonomy_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2reportsTaxonomyNamesDataReportPage
    TaxonomyNamesPost = function(v2_taxonomy_metadata_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyNamesPostWithHttpInfo(v2_taxonomy_metadata_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifiers to get taxonomic names data report by post
    #'
    #' @param v2_taxonomy_metadata_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2reportsTaxonomyNamesDataReportPage) with additional information such as HTTP status code, headers
    TaxonomyNamesPostWithHttpInfo = function(v2_taxonomy_metadata_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_metadata_request`)) {
        stop("Missing required parameter `v2_taxonomy_metadata_request`.")
      }


      if (!is.null(`v2_taxonomy_metadata_request`)) {
        local_var_body <- `v2_taxonomy_metadata_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/name_report"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2reportsTaxonomyNamesDataReportPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifier to get related taxonomic identifiers, such as children
    #'
    #' @param tax_id 
    #' @param include_lineage (optional) If true, return reports for all taxonomy nodes in the lineages of the requested tax_id (default value: FALSE)
    #' @param include_subtree (optional) This field is deprecated because all requests include the subtree, so it has no effect (default value: FALSE)
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param page_size (optional) The maximum number of taxids to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from a `GetRelatedTaxids` call with more than `page_size` results. Use this token, along with the previous `TaxonomyRelatedIdRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyTaxIdsPage
    TaxonomyRelatedIds = function(tax_id, include_lineage = FALSE, include_subtree = FALSE, ranks = NULL, page_size = 20, page_token = NULL, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyRelatedIdsWithHttpInfo(tax_id, include_lineage, include_subtree, ranks, page_size, page_token, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifier to get related taxonomic identifiers, such as children
    #'
    #' @param tax_id 
    #' @param include_lineage (optional) If true, return reports for all taxonomy nodes in the lineages of the requested tax_id (default value: FALSE)
    #' @param include_subtree (optional) This field is deprecated because all requests include the subtree, so it has no effect (default value: FALSE)
    #' @param ranks (optional) Only include taxons of the provided ranks. If empty, return all ranks.
    #' @param page_size (optional) The maximum number of taxids to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results. (default value: 20)
    #' @param page_token (optional) A page token is returned from a `GetRelatedTaxids` call with more than `page_size` results. Use this token, along with the previous `TaxonomyRelatedIdRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyTaxIdsPage) with additional information such as HTTP status code, headers
    TaxonomyRelatedIdsWithHttpInfo = function(tax_id, include_lineage = FALSE, include_subtree = FALSE, ranks = NULL, page_size = 20, page_token = NULL, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`tax_id`)) {
        stop("Missing required parameter `tax_id`.")
      }







      query_params[["include_lineage"]] <- `include_lineage`

      query_params[["include_subtree"]] <- `include_subtree`

      # explore
      for (query_item in `ranks`) {
        query_params[["ranks"]] <- c(query_params[["ranks"]], list(`ranks` = query_item))
      }

      query_params[["page_size"]] <- `page_size`

      query_params[["page_token"]] <- `page_token`

      local_var_url_path <- "/taxonomy/taxon/{tax_id}/related_ids"
      if (!missing(`tax_id`)) {
        local_var_url_path <- gsub("\\{tax_id\\}", URLencode(as.character(`tax_id`), reserved = TRUE), local_var_url_path)
      }

      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list()

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "GET",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyTaxIdsPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    },

    #' @description
    #' Use taxonomic identifier to get related taxonomic identifiers, such as children
    #'
    #' @param v2_taxonomy_related_id_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return V2TaxonomyTaxIdsPage
    TaxonomyRelatedIdsPost = function(v2_taxonomy_related_id_request, data_file = NULL, ...) {
      local_var_response <- self$TaxonomyRelatedIdsPostWithHttpInfo(v2_taxonomy_related_id_request, data_file = data_file, ...)
      if (local_var_response$status_code >= 200 && local_var_response$status_code <= 299) {
        return(local_var_response$content)
      } else if (local_var_response$status_code >= 300 && local_var_response$status_code <= 399) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 400 && local_var_response$status_code <= 499) {
        return(local_var_response)
      } else if (local_var_response$status_code >= 500 && local_var_response$status_code <= 599) {
        return(local_var_response)
      }
    },

    #' @description
    #' Use taxonomic identifier to get related taxonomic identifiers, such as children
    #'
    #' @param v2_taxonomy_related_id_request 
    #' @param data_file (optional) name of the data file to save the result
    #' @param ... Other optional arguments
    #'
    #' @return API response (V2TaxonomyTaxIdsPage) with additional information such as HTTP status code, headers
    TaxonomyRelatedIdsPostWithHttpInfo = function(v2_taxonomy_related_id_request, data_file = NULL, ...) {
      args <- list(...)
      query_params <- list()
      header_params <- c()
      form_params <- list()
      file_params <- list()
      local_var_body <- NULL
      oauth_scopes <- NULL
      is_oauth <- FALSE

      if (missing(`v2_taxonomy_related_id_request`)) {
        stop("Missing required parameter `v2_taxonomy_related_id_request`.")
      }


      if (!is.null(`v2_taxonomy_related_id_request`)) {
        local_var_body <- `v2_taxonomy_related_id_request`$toJSONString()
      } else {
        local_var_body <- NULL
      }

      local_var_url_path <- "/taxonomy/related_ids"
      # API key authentication
      if ("api-key" %in% names(self$api_client$api_keys) && nchar(self$api_client$api_keys["api-key"]) > 0) {
        header_params["api-key"] <- paste(unlist(self$api_client$api_keys["api-key"]), collapse = "")
      }

      # The Accept request HTTP header
      local_var_accepts <- list("text/plain", "application/json")

      # The Content-Type representation header
      local_var_content_types <- list("application/json")

      local_var_resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, local_var_url_path),
                                 method = "POST",
                                 query_params = query_params,
                                 header_params = header_params,
                                 form_params = form_params,
                                 file_params = file_params,
                                 accepts = local_var_accepts,
                                 content_types = local_var_content_types,
                                 body = local_var_body,
                                 is_oauth = is_oauth,
                                 oauth_scopes = oauth_scopes,
                                 ...)

      if (local_var_resp$status_code >= 200 && local_var_resp$status_code <= 299) {
        # save response in a file
        if (!is.null(data_file)) {
          self$api_client$WriteFile(local_var_resp, data_file)
        }

        deserialized_resp_obj <- tryCatch(
          self$api_client$DeserializeResponse(local_var_resp, "V2TaxonomyTaxIdsPage"),
          error = function(e) {
            stop("Failed to deserialize response")
          }
        )
        local_var_resp$content <- deserialized_resp_obj
        return(local_var_resp)
      } 
      
      local_var_error_msg <- local_var_resp$response_as_text()      
      if (local_var_resp$status_code >= 300 && local_var_resp$status_code <= 399) {
        ApiResponse$new(paste("Server returned ", local_var_resp$status_code, " response status code."), local_var_resp)
      } else if (local_var_resp$status_code >= 400 && local_var_resp$status_code <= 499) {
        ApiResponse$new("API client error", local_var_resp)
      } else if (local_var_resp$status_code >= 500 && local_var_resp$status_code <= 599) {
        if (is.null(local_var_resp$response) || local_var_resp$response == "") {
          local_var_resp$response <- "API server error"
        }
        return(local_var_resp)
      }
    }
  )
)
